{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    },
    "toc": "true"
   },
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1 toc-item\"><a href=\"#TSS\" data-toc-modified-id=\"TSS-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>TSS</a></div><div class=\"lev1 toc-item\"><a href=\"#tRNA\" data-toc-modified-id=\"tRNA-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>tRNA</a></div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import warnings\n",
    "import logging\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import copy\n",
    "import re\n",
    "from gffutils.pybedtools_integration import tsses\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "gencode_vM11 = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.annotation.gtf'\n",
    "gencode_vM11_db = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.annotation.gtf.db'\n",
    "prefix = '/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils'\n",
    "old_prefix = '/home/cmb-panasas2/skchoudh/github_projects/gencode_regions/data/GRCh37/vM11/'\n",
    "mm10_chrsizes = '/home/cmb-panasas2/skchoudh/genomes/mm10/fasta/mm10.chrom.sizes'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "#db = gffutils.create_db(gencode_vM11, dbfn=gencode_vM11_db, merge_strategy='merge',\n",
    "#                        disable_infer_genes=True, disable_infer_transcripts=True, force=True)\n",
    "def create_gene_dict(db):\n",
    "    '''\n",
    "    Store each feature line db.all_features() as a dict of dicts\n",
    "    '''\n",
    "    gene_dict = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))\n",
    "    for line_no, feature in enumerate(db.all_features()):\n",
    "        gene_ids = feature.attributes['gene_id']\n",
    "        feature_type = feature.featuretype\n",
    "        if feature_type == 'gene':\n",
    "            if len(gene_ids)!=1:\n",
    "                logging.warning('Found multiple gene_ids on line {} in gtf'.format(line_no))\n",
    "                break\n",
    "            else:\n",
    "                gene_id = gene_ids[0]\n",
    "                gene_dict[gene_id]['gene'] = feature\n",
    "        else:\n",
    "            transcript_ids = feature.attributes['transcript_id']\n",
    "\n",
    "            for gene_id in gene_ids:\n",
    "                for transcript_id in transcript_ids:\n",
    "                    gene_dict[gene_id][transcript_id][feature_type].append(feature)\n",
    "    return gene_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(gencode_vM11_db, keep_order=True)\n",
    "gene_dict = create_gene_dict(db)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "def get_gene_list(gene_dict):\n",
    "    return list(set(gene_dict.keys()))\n",
    "\n",
    "def get_UTR_regions(gene_dict, gene_id, transcript, cds):\n",
    "    if len(cds)==0:\n",
    "        return [], []\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    utrs = gene_dict[gene_id][transcript]['UTR']\n",
    "    first_cds = cds[0]\n",
    "    last_cds = cds[-1]\n",
    "    for utr in utrs:\n",
    "        ## Push all cds at once\n",
    "        ## Sort later to remove duplicates\n",
    "        strand = utr.strand\n",
    "        if strand == '+':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')\n",
    "        elif strand == '-':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)                \n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')    \n",
    "    return utr5_regions, utr3_regions\n",
    "    \n",
    "def create_bed(regions, bedtype='0'):\n",
    "    '''Create bed from list of regions\n",
    "    bedtype: 0 or 1\n",
    "        0-Based or 1-based coordinate of the BED\n",
    "    '''\n",
    "    bedstr = ''\n",
    "    for region in regions:\n",
    "        assert len(region.attributes['gene_id']) == 1\n",
    "        ## GTF start is 1-based, so shift by one while writing \n",
    "        ## to 0-based BED format\n",
    "        if bedtype == '0':\n",
    "            start = region.start - 1\n",
    "        else:\n",
    "            start = region.start\n",
    "        bedstr += '{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(region.chrom,\n",
    "                                             start,\n",
    "                                             region.stop,\n",
    "                                             re.sub('\\.\\d+', '', region.attributes['gene_id'][0]),\n",
    "                                             '.',\n",
    "                                             region.strand)\n",
    "    return bedstr\n",
    "\n",
    "def rename_regions(regions, gene_id):\n",
    "    regions = list(regions)\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    for region in regions:\n",
    "        region.attributes['gene_id'] = gene_id\n",
    "    return regions\n",
    "\n",
    "def merge_regions(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start))\n",
    "    return merged\n",
    "\n",
    "def merge_regions_nostrand(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start), ignore_strand=True)\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.cds.bed)>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_bed = ''\n",
    "utr3_bed = ''\n",
    "gene_bed = ''\n",
    "exon_bed = ''\n",
    "intron_bed = ''\n",
    "start_codon_bed = ''\n",
    "stop_codon_bed = ''\n",
    "cds_bed = ''\n",
    "\n",
    "gene_list = []\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    gene_list.append(gene_dict[gene_id]['gene'])\n",
    "    \n",
    "    utr5_regions, utr3_regions = [], []\n",
    "    exon_regions, intron_regions = [], []\n",
    "    star_codon_regions, stop_codon_regions = [], []\n",
    "    cds_regions = []\n",
    "    \n",
    "    for feature in gene_dict[gene_id].keys():\n",
    "        if feature == 'gene':\n",
    "            continue\n",
    "        cds = list(gene_dict[gene_id][feature]['CDS'])\n",
    "        exons = list(gene_dict[gene_id][feature]['exon'])\n",
    "        merged_exons = merge_regions(db, exons)\n",
    "        introns = db.interfeatures(merged_exons)\n",
    "        utr5_region, utr3_region = get_UTR_regions(gene_dict, gene_id, feature, cds)\n",
    "        utr5_regions += utr5_region\n",
    "        utr3_regions += utr3_region\n",
    "        exon_regions += exons\n",
    "        intron_regions += introns\n",
    "        cds_regions += cds\n",
    "        \n",
    "    merged_utr5 = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5 = rename_regions(merged_utr5, gene_id)\n",
    "    \n",
    "    merged_utr3 = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3 = rename_regions(merged_utr3, gene_id)\n",
    "    \n",
    "    merged_exons = merge_regions(db, exon_regions)\n",
    "    renamed_exons = rename_regions(merged_exons, gene_id)\n",
    "    \n",
    "    merged_introns = merge_regions(db, intron_regions)\n",
    "    renamed_introns = rename_regions(merged_introns, gene_id)\n",
    "    \n",
    "    merged_cds = merge_regions(db, cds_regions)\n",
    "    renamed_cds = rename_regions(merged_cds, gene_id)\n",
    "    \n",
    "    utr3_bed += create_bed(renamed_utr3)\n",
    "    utr5_bed += create_bed(renamed_utr5)\n",
    "    exon_bed += create_bed(renamed_exons)\n",
    "    intron_bed += create_bed(renamed_introns)\n",
    "    cds_bed += create_bed(renamed_cds)\n",
    "    \n",
    "    \n",
    "gene_bed = create_bed(gene_list)\n",
    "gene_bedtool = pybedtools.BedTool(gene_bed, from_string=True)\n",
    "utr5_bedtool = pybedtools.BedTool(utr5_bed, from_string=True)\n",
    "utr3_bedtool = pybedtools.BedTool(utr3_bed, from_string=True)\n",
    "exon_bedtool = pybedtools.BedTool(exon_bed, from_string=True)\n",
    "intron_bedtool = pybedtools.BedTool(intron_bed, from_string=True)\n",
    "cds_bedtool = pybedtools.BedTool(cds_bed, from_string=True)\n",
    "\n",
    "gene_bedtool.remove_invalid().sort().saveas('{}.genes.bed'.format(prefix))\n",
    "utr5_bedtool.remove_invalid().sort().saveas('{}.UTR5.bed'.format(prefix))\n",
    "utr3_bedtool.remove_invalid().sort().saveas('{}.UTR3.bed'.format(prefix))\n",
    "exon_bedtool.remove_invalid().sort().saveas('{}.exon.bed'.format(prefix))\n",
    "intron_bedtool.remove_invalid().sort().saveas('{}.intron.bed'.format(prefix))\n",
    "cds_bedtool.remove_invalid().sort().saveas('{}.cds.bed'.format(prefix))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.stop_codon.bed)>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for gene_id in get_gene_list(gene_dict):\n",
    "    start_codons = []\n",
    "    stop_codons = []\n",
    "    for start_codon in db.children(gene_id, featuretype='start_codon'):\n",
    "        ## 1 -based stop\n",
    "        ## 0-based start handled while converting to bed\n",
    "        start_codon.stop = start_codon.start\n",
    "        start_codons.append(start_codon)\n",
    "    for stop_codon in db.children(gene_id, featuretype='stop_codon'):\n",
    "        stop_codon.start = stop_codon.stop\n",
    "        stop_codon.stop = stop_codon.stop+1\n",
    "        stop_codons.append(stop_codon)\n",
    "    merged_start_codons = merge_regions(db, start_codons)\n",
    "    renamed_start_codons = rename_regions(merged_start_codons, gene_id)\n",
    "    merged_stop_codons = merge_regions(db, stop_codons)\n",
    "    renamed_stop_codons = rename_regions(merged_stop_codons, gene_id)\n",
    "    \n",
    "    start_codon_bed += create_bed(renamed_start_codons)    \n",
    "    stop_codon_bed += create_bed(renamed_stop_codons)\n",
    "\n",
    "    \n",
    "start_codon_bedtool = pybedtools.BedTool(start_codon_bed, from_string=True)\n",
    "stop_codon_bedtool = pybedtools.BedTool(stop_codon_bed, from_string=True)\n",
    "start_codon_bedtool.remove_invalid().sort().saveas('{}.start_codon.bed'.format(prefix))\n",
    "stop_codon_bedtool.remove_invalid().sort().saveas('{}.stop_codon.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.tss_sites.bed)>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TSS\n",
    "polyA_sites_bed = ''\n",
    "tss_sites_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    tss_sites = []\n",
    "    polyA_sites = []\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        start_t = copy.deepcopy(transcript)\n",
    "        stop_t = copy.deepcopy(transcript)\n",
    "        \n",
    "        start_t.stop = start_t.start + 1\n",
    "        \n",
    "        stop_t.start = stop_t.stop\n",
    "        \n",
    "        if transcript.strand == '-':\n",
    "            start_t, stop_t = stop_t, start_t\n",
    "        polyA_sites.append(start_t)\n",
    "        tss_sites.append(stop_t)\n",
    "    merged_polyA_sites = merge_regions(db, polyA_sites)\n",
    "    renamed_polyA_sites = rename_regions(merged_polyA_sites, gene_id)    \n",
    "    merged_tss_sites = merge_regions(db, tss_sites)\n",
    "    renamed_tss_sites = rename_regions(merged_tss_sites, gene_id)\n",
    "    polyA_sites_bed += create_bed(renamed_polyA_sites)    \n",
    "    tss_sites_bed += create_bed(renamed_tss_sites)\n",
    "\n",
    "polyA_sites_bedtool = pybedtools.BedTool(polyA_sites_bed, from_string=True)\n",
    "tss_sites_bedtool = pybedtools.BedTool(tss_sites_bed, from_string=True)\n",
    "polyA_sites_bedtool.remove_invalid().sort().saveas('{}.polyA_sites.bed'.format(prefix))\n",
    "tss_sites_bedtool.remove_invalid().sort().saveas('{}.tss_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "source": [
    "# TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "tss = tsses(db, as_bed6=True, merge_overlapping=True)\n",
    "tss.remove_invalid().sort().saveas('{}.tss_temp.bed'.format(prefix))\n",
    "promoter = tss.slop(l=1000, r=1000, s=True, g=mm10_chrsizes)\n",
    "promoter.remove_invalid().sort().saveas('{}.promoter.1000.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "source": [
    "# tRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.tRNA_sites.bed)>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites = []\n",
    "tRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'tRNA' in transcript.attributes['gene_type'] or 'Mt_tRNA' in transcript.attributes['gene_type']:\n",
    "            tRNA_sites.append(transcript)\n",
    "    #merged_tRNA_sites = merge_regions_nostrand(db, tRNA_sites)\n",
    "    #renamed_tRNA_sites = rename_regions(merged_tRNA_sites, gene_id)\n",
    "    tRNA_bed += create_bed(tRNA_sites)\n",
    "\n",
    "tRNA_sites_bedtool = pybedtools.BedTool(tRNA_bed, from_string=True)\n",
    "tRNA_sites_bedtool.remove_invalid().sort().saveas('{}.tRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000064366</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000086508</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000000791</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000028266</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000077703</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000022941</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000037902</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000079103</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000095864</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000087898</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000018593</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000100817</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000029714</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000036211</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000028933</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000090334</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000074115</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000049894</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000081782</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000076552</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000088385</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000085768</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000022452</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000036907</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000051095</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000049018</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000110108</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000083054</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000046790</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11670</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000109446</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463140</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5017</td>\n",
       "      <td>5086</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463141</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5088</td>\n",
       "      <td>5159</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463142</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5191</td>\n",
       "      <td>5257</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463143</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5259</td>\n",
       "      <td>5326</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463144</th>\n",
       "      <td>chrM</td>\n",
       "      <td>6869</td>\n",
       "      <td>6938</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463145</th>\n",
       "      <td>chrM</td>\n",
       "      <td>6941</td>\n",
       "      <td>7011</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463146</th>\n",
       "      <td>chrM</td>\n",
       "      <td>7699</td>\n",
       "      <td>7764</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463147</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9390</td>\n",
       "      <td>9458</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463148</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9807</td>\n",
       "      <td>9875</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463149</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11545</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463150</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14070</td>\n",
       "      <td>14139</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463151</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15288</td>\n",
       "      <td>15422</td>\n",
       "      <td>ENSMUSG00000095433</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463152</th>\n",
       "      <td>chrM</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463153</th>\n",
       "      <td>chrM</td>\n",
       "      <td>1024</td>\n",
       "      <td>1093</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463154</th>\n",
       "      <td>chrM</td>\n",
       "      <td>2675</td>\n",
       "      <td>2750</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463155</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3705</td>\n",
       "      <td>3842</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463156</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3844</td>\n",
       "      <td>3913</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463157</th>\n",
       "      <td>chrM</td>\n",
       "      <td>4949</td>\n",
       "      <td>5016</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463158</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5017</td>\n",
       "      <td>5086</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463159</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5088</td>\n",
       "      <td>5159</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463160</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5191</td>\n",
       "      <td>5257</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463161</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5259</td>\n",
       "      <td>5326</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463162</th>\n",
       "      <td>chrM</td>\n",
       "      <td>6869</td>\n",
       "      <td>6938</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463163</th>\n",
       "      <td>chrM</td>\n",
       "      <td>6941</td>\n",
       "      <td>7011</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463164</th>\n",
       "      <td>chrM</td>\n",
       "      <td>7699</td>\n",
       "      <td>7764</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463165</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9390</td>\n",
       "      <td>9458</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463166</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9807</td>\n",
       "      <td>9875</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463167</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11545</td>\n",
       "      <td>11741</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463168</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14070</td>\n",
       "      <td>14139</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463169</th>\n",
       "      <td>chrM</td>\n",
       "      <td>15288</td>\n",
       "      <td>15422</td>\n",
       "      <td>ENSMUSG00000092425</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>463170 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  start    end                name score strand\n",
       "0       chrM  11670  11741  ENSMUSG00000064366     .      .\n",
       "1       chrM  11670  11741  ENSMUSG00000086508     .      .\n",
       "2       chrM  11670  11741  ENSMUSG00000000791     .      .\n",
       "3       chrM  11670  11741  ENSMUSG00000028266     .      .\n",
       "4       chrM  11670  11741  ENSMUSG00000077703     .      .\n",
       "5       chrM  11670  11741  ENSMUSG00000022941     .      .\n",
       "6       chrM  11670  11741  ENSMUSG00000037902     .      .\n",
       "7       chrM  11670  11741  ENSMUSG00000079103     .      .\n",
       "8       chrM  11670  11741  ENSMUSG00000095864     .      .\n",
       "9       chrM  11670  11741  ENSMUSG00000087898     .      .\n",
       "10      chrM  11670  11741  ENSMUSG00000018593     .      .\n",
       "11      chrM  11670  11741  ENSMUSG00000100817     .      .\n",
       "12      chrM  11670  11741  ENSMUSG00000029714     .      .\n",
       "13      chrM  11670  11741  ENSMUSG00000036211     .      .\n",
       "14      chrM  11670  11741  ENSMUSG00000028933     .      .\n",
       "15      chrM  11670  11741  ENSMUSG00000090334     .      .\n",
       "16      chrM  11670  11741  ENSMUSG00000074115     .      .\n",
       "17      chrM  11670  11741  ENSMUSG00000049894     .      .\n",
       "18      chrM  11670  11741  ENSMUSG00000081782     .      .\n",
       "19      chrM  11670  11741  ENSMUSG00000076552     .      .\n",
       "20      chrM  11670  11741  ENSMUSG00000088385     .      .\n",
       "21      chrM  11670  11741  ENSMUSG00000085768     .      .\n",
       "22      chrM  11670  11741  ENSMUSG00000022452     .      .\n",
       "23      chrM  11670  11741  ENSMUSG00000036907     .      .\n",
       "24      chrM  11670  11741  ENSMUSG00000051095     .      .\n",
       "25      chrM  11670  11741  ENSMUSG00000049018     .      .\n",
       "26      chrM  11670  11741  ENSMUSG00000110108     .      .\n",
       "27      chrM  11670  11741  ENSMUSG00000083054     .      .\n",
       "28      chrM  11670  11741  ENSMUSG00000046790     .      .\n",
       "29      chrM  11670  11741  ENSMUSG00000109446     .      .\n",
       "...      ...    ...    ...                 ...   ...    ...\n",
       "463140  chrM   5017   5086  ENSMUSG00000095433     .      .\n",
       "463141  chrM   5088   5159  ENSMUSG00000095433     .      .\n",
       "463142  chrM   5191   5257  ENSMUSG00000095433     .      .\n",
       "463143  chrM   5259   5326  ENSMUSG00000095433     .      .\n",
       "463144  chrM   6869   6938  ENSMUSG00000095433     .      .\n",
       "463145  chrM   6941   7011  ENSMUSG00000095433     .      .\n",
       "463146  chrM   7699   7764  ENSMUSG00000095433     .      .\n",
       "463147  chrM   9390   9458  ENSMUSG00000095433     .      .\n",
       "463148  chrM   9807   9875  ENSMUSG00000095433     .      .\n",
       "463149  chrM  11545  11741  ENSMUSG00000095433     .      .\n",
       "463150  chrM  14070  14139  ENSMUSG00000095433     .      .\n",
       "463151  chrM  15288  15422  ENSMUSG00000095433     .      .\n",
       "463152  chrM      0     68  ENSMUSG00000092425     .      .\n",
       "463153  chrM   1024   1093  ENSMUSG00000092425     .      .\n",
       "463154  chrM   2675   2750  ENSMUSG00000092425     .      .\n",
       "463155  chrM   3705   3842  ENSMUSG00000092425     .      .\n",
       "463156  chrM   3844   3913  ENSMUSG00000092425     .      .\n",
       "463157  chrM   4949   5016  ENSMUSG00000092425     .      .\n",
       "463158  chrM   5017   5086  ENSMUSG00000092425     .      .\n",
       "463159  chrM   5088   5159  ENSMUSG00000092425     .      .\n",
       "463160  chrM   5191   5257  ENSMUSG00000092425     .      .\n",
       "463161  chrM   5259   5326  ENSMUSG00000092425     .      .\n",
       "463162  chrM   6869   6938  ENSMUSG00000092425     .      .\n",
       "463163  chrM   6941   7011  ENSMUSG00000092425     .      .\n",
       "463164  chrM   7699   7764  ENSMUSG00000092425     .      .\n",
       "463165  chrM   9390   9458  ENSMUSG00000092425     .      .\n",
       "463166  chrM   9807   9875  ENSMUSG00000092425     .      .\n",
       "463167  chrM  11545  11741  ENSMUSG00000092425     .      .\n",
       "463168  chrM  14070  14139  ENSMUSG00000092425     .      .\n",
       "463169  chrM  15288  15422  ENSMUSG00000092425     .      .\n",
       "\n",
       "[463170 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/mm10/annotation/gencode.vM11.gffutils.rRNA_sites.bed)>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites = []\n",
    "rRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'rRNA' in transcript.attributes['gene_type']:\n",
    "            rRNA_sites.append(transcript)\n",
    "    #renamed_rRNA_sites = rename_regions(rRNA_sites, gene_id)\n",
    "    rRNA_bed += create_bed(rRNA_sites)\n",
    "\n",
    "rRNA_sites_bedtool = pybedtools.BedTool(rRNA_bed, from_string=True)\n",
    "rRNA_sites_bedtool.remove_invalid().sort().saveas('{}.rRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "12px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": true,
   "toc_section_display": "block",
   "toc_window_display": true,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
