{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, OrderedDict\n",
    "import warnings\n",
    "import logging\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import copy\n",
    "import re\n",
    "from gffutils.pybedtools_integration import tsses\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gtf'\n",
    "gtf_db = '/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gtf.db'\n",
    "prefix = '/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils'\n",
    "chrsizes = '/home/cmb-panasas2/skchoudh/genomes/WBcel235/fasta/Caenorhabditis_elegans.WBcel235.dna.toplevel.sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.create_db(gtf, dbfn=gtf_db, disable_infer_genes=True, disable_infer_transcripts=True, merge_strategy='merge', force=True)\n",
    "def create_gene_dict(db):\n",
    "    '''\n",
    "    Store each feature line db.all_features() as a dict of dicts\n",
    "    '''\n",
    "    gene_dict = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))\n",
    "    for line_no, feature in enumerate(db.all_features()):\n",
    "        gene_ids = feature.attributes['gene_id']\n",
    "        feature_type = feature.featuretype\n",
    "        if feature_type == 'gene':\n",
    "            if len(gene_ids)!=1:\n",
    "                logging.warning('Found multiple gene_ids on line {} in gtf'.format(line_no))\n",
    "                break\n",
    "            else:\n",
    "                gene_id = gene_ids[0]\n",
    "                gene_dict[gene_id]['gene'] = feature\n",
    "        else:\n",
    "            transcript_ids = feature.attributes['transcript_id']\n",
    "\n",
    "            for gene_id in gene_ids:\n",
    "                for transcript_id in transcript_ids:\n",
    "                    gene_dict[gene_id][transcript_id][feature_type].append(feature)\n",
    "    return gene_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(gtf_db, keep_order=True)\n",
    "gene_dict = create_gene_dict(db)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS\n",
      "Selenocysteine\n",
      "exon\n",
      "five_prime_utr\n",
      "gene\n",
      "start_codon\n",
      "stop_codon\n",
      "three_prime_utr\n",
      "transcript\n"
     ]
    }
   ],
   "source": [
    "for x in db.featuretypes():\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_list(gene_dict):\n",
    "    return list(set(gene_dict.keys()))\n",
    "\n",
    "def get_UTR_regions(gene_dict, gene_id, transcript, cds):\n",
    "    if len(cds)==0:\n",
    "        return [], []\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    utrs = gene_dict[gene_id][transcript]['UTR']\n",
    "    first_cds = cds[0]\n",
    "    last_cds = cds[-1]\n",
    "    for utr in utrs:\n",
    "        ## Push all cds at once\n",
    "        ## Sort later to remove duplicates\n",
    "        strand = utr.strand\n",
    "        if strand == '+':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')\n",
    "        elif strand == '-':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)                \n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')    \n",
    "    return utr5_regions, utr3_regions\n",
    "    \n",
    "def create_bed(regions, bedtype='0'):\n",
    "    '''Create bed from list of regions\n",
    "    bedtype: 0 or 1\n",
    "        0-Based or 1-based coordinate of the BED\n",
    "    '''\n",
    "    bedstr = ''\n",
    "    for region in regions:\n",
    "        assert len(region.attributes['gene_id']) == 1\n",
    "        ## GTF start is 1-based, so shift by one while writing \n",
    "        ## to 0-based BED format\n",
    "        if bedtype == '0':\n",
    "            start = region.start - 1\n",
    "        else:\n",
    "            start = region.start\n",
    "        bedstr += '{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(region.chrom,\n",
    "                                             start,\n",
    "                                             region.stop,\n",
    "                                             re.sub('\\.\\d+', '', region.attributes['gene_id'][0]),\n",
    "                                             '.',\n",
    "                                             region.strand)\n",
    "    return bedstr\n",
    "\n",
    "def rename_regions(regions, gene_id):\n",
    "    regions = list(regions)\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    for region in regions:\n",
    "        region.attributes['gene_id'] = gene_id\n",
    "    return regions\n",
    "\n",
    "def merge_regions(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start))\n",
    "    return merged\n",
    "\n",
    "def merge_regions_nostrand(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start), ignore_strand=True)\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils.cds.bed)>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_bed = ''\n",
    "utr3_bed = ''\n",
    "gene_bed = ''\n",
    "exon_bed = ''\n",
    "intron_bed = ''\n",
    "start_codon_bed = ''\n",
    "stop_codon_bed = ''\n",
    "cds_bed = ''\n",
    "\n",
    "gene_list = []\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    gene_list.append(gene_dict[gene_id]['gene'])\n",
    "    \n",
    "    utr5_regions, utr3_regions = [], []\n",
    "    exon_regions, intron_regions = [], []\n",
    "    star_codon_regions, stop_codon_regions = [], []\n",
    "    cds_regions = []\n",
    "    \n",
    "    for feature in gene_dict[gene_id].keys():\n",
    "        if feature == 'gene':\n",
    "            continue\n",
    "        cds = list(gene_dict[gene_id][feature]['CDS'])\n",
    "        exons = list(gene_dict[gene_id][feature]['exon'])\n",
    "        merged_exons = merge_regions(db, exons)\n",
    "        introns = db.interfeatures(merged_exons)\n",
    "        #utr5_region, utr3_region = get_UTR_regions(gene_dict, gene_id, feature, cds)\n",
    "        utr5_region = list(gene_dict[gene_id][feature]['five_prime_utr'])\n",
    "        utr3_region = list(gene_dict[gene_id][feature]['three_prime_utr'])\n",
    "        utr5_regions += utr5_region\n",
    "        utr3_regions += utr3_region\n",
    "        exon_regions += exons\n",
    "        intron_regions += introns\n",
    "        cds_regions += cds\n",
    "        \n",
    "    merged_utr5 = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5 = rename_regions(merged_utr5, gene_id)\n",
    "    \n",
    "    merged_utr3 = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3 = rename_regions(merged_utr3, gene_id)\n",
    "    \n",
    "    merged_exons = merge_regions(db, exon_regions)\n",
    "    renamed_exons = rename_regions(merged_exons, gene_id)\n",
    "    \n",
    "    merged_introns = merge_regions(db, intron_regions)\n",
    "    renamed_introns = rename_regions(merged_introns, gene_id)\n",
    "    \n",
    "    merged_cds = merge_regions(db, cds_regions)\n",
    "    renamed_cds = rename_regions(merged_cds, gene_id)\n",
    "    \n",
    "    utr3_bed += create_bed(renamed_utr3)\n",
    "    utr5_bed += create_bed(renamed_utr5)\n",
    "    exon_bed += create_bed(renamed_exons)\n",
    "    intron_bed += create_bed(renamed_introns)\n",
    "    cds_bed += create_bed(renamed_cds)\n",
    "    \n",
    "    \n",
    "gene_bed = create_bed(gene_list)\n",
    "gene_bedtool = pybedtools.BedTool(gene_bed, from_string=True)\n",
    "utr5_bedtool = pybedtools.BedTool(utr5_bed, from_string=True)\n",
    "utr3_bedtool = pybedtools.BedTool(utr3_bed, from_string=True)\n",
    "exon_bedtool = pybedtools.BedTool(exon_bed, from_string=True)\n",
    "intron_bedtool = pybedtools.BedTool(intron_bed, from_string=True)\n",
    "cds_bedtool = pybedtools.BedTool(cds_bed, from_string=True)\n",
    "\n",
    "gene_bedtool.remove_invalid().sort().saveas('{}.genes.bed'.format(prefix))\n",
    "utr5_bedtool.remove_invalid().sort().saveas('{}.UTR5.bed'.format(prefix))\n",
    "utr3_bedtool.remove_invalid().sort().saveas('{}.UTR3.bed'.format(prefix))\n",
    "exon_bedtool.remove_invalid().sort().saveas('{}.exon.bed'.format(prefix))\n",
    "intron_bedtool.remove_invalid().sort().saveas('{}.intron.bed'.format(prefix))\n",
    "cds_bedtool.remove_invalid().sort().saveas('{}.cds.bed'.format(prefix))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils.stop_codon.bed)>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for gene_id in get_gene_list(gene_dict):\n",
    "    start_codons = []\n",
    "    stop_codons = []\n",
    "    for start_codon in db.children(gene_id, featuretype='start_codon'):\n",
    "        ## 1 -based stop\n",
    "        ## 0-based start handled while converting to bed\n",
    "        start_codon.stop = start_codon.start\n",
    "        start_codons.append(start_codon)\n",
    "    for stop_codon in db.children(gene_id, featuretype='stop_codon'):\n",
    "        stop_codon.start = stop_codon.stop\n",
    "        stop_codon.stop = stop_codon.stop+1\n",
    "        stop_codons.append(stop_codon)\n",
    "    merged_start_codons = merge_regions(db, start_codons)\n",
    "    renamed_start_codons = rename_regions(merged_start_codons, gene_id)\n",
    "    merged_stop_codons = merge_regions(db, stop_codons)\n",
    "    renamed_stop_codons = rename_regions(merged_stop_codons, gene_id)\n",
    "    \n",
    "    start_codon_bed += create_bed(renamed_start_codons)    \n",
    "    stop_codon_bed += create_bed(renamed_stop_codons)\n",
    "\n",
    "    \n",
    "start_codon_bedtool = pybedtools.BedTool(start_codon_bed, from_string=True)\n",
    "stop_codon_bedtool = pybedtools.BedTool(stop_codon_bed, from_string=True)\n",
    "start_codon_bedtool.remove_invalid().sort().saveas('{}.start_codon.bed'.format(prefix))\n",
    "stop_codon_bedtool.remove_invalid().sort().saveas('{}.stop_codon.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils.tss_sites.bed)>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TSS\n",
    "polyA_sites_bed = ''\n",
    "tss_sites_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    tss_sites = []\n",
    "    polyA_sites = []\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        start_t = copy.deepcopy(transcript)\n",
    "        stop_t = copy.deepcopy(transcript)\n",
    "        \n",
    "        start_t.stop = start_t.start + 1\n",
    "        \n",
    "        stop_t.start = stop_t.stop\n",
    "        \n",
    "        if transcript.strand == '-':\n",
    "            start_t, stop_t = stop_t, start_t\n",
    "        polyA_sites.append(start_t)\n",
    "        tss_sites.append(stop_t)\n",
    "    merged_polyA_sites = merge_regions(db, polyA_sites)\n",
    "    renamed_polyA_sites = rename_regions(merged_polyA_sites, gene_id)    \n",
    "    merged_tss_sites = merge_regions(db, tss_sites)\n",
    "    renamed_tss_sites = rename_regions(merged_tss_sites, gene_id)\n",
    "    polyA_sites_bed += create_bed(renamed_polyA_sites)    \n",
    "    tss_sites_bed += create_bed(renamed_tss_sites)\n",
    "\n",
    "polyA_sites_bedtool = pybedtools.BedTool(polyA_sites_bed, from_string=True)\n",
    "tss_sites_bedtool = pybedtools.BedTool(tss_sites_bed, from_string=True)\n",
    "polyA_sites_bedtool.remove_invalid().sort().saveas('{}.polyA_sites.bed'.format(prefix))\n",
    "tss_sites_bedtool.remove_invalid().sort().saveas('{}.tss_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils.tRNA_sites.bed)>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites = []\n",
    "tRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'tRNA' in transcript.attributes['gene_biotype'] or 'Mt_tRNA' in transcript.attributes['transcript_biotype']:\n",
    "            tRNA_sites.append(transcript)\n",
    "    #merged_tRNA_sites = merge_regions_nostrand(db, tRNA_sites)\n",
    "    #renamed_tRNA_sites = rename_regions(merged_tRNA_sites, gene_id)\n",
    "    tRNA_bed += create_bed(tRNA_sites)\n",
    "\n",
    "tRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(tRNA_bed.split('\\n'))))\n",
    "\n",
    "tRNA_sites_bedtool = pybedtools.BedTool(tRNA_bed, from_string=True)\n",
    "tRNA_sites_bedtool.remove_invalid().sort().saveas('{}.tRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X</td>\n",
       "      <td>8409295</td>\n",
       "      <td>8409377</td>\n",
       "      <td>WBGene00022981</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X</td>\n",
       "      <td>8417911</td>\n",
       "      <td>8417993</td>\n",
       "      <td>WBGene00022980</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X</td>\n",
       "      <td>8820184</td>\n",
       "      <td>8820256</td>\n",
       "      <td>WBGene00023119</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X</td>\n",
       "      <td>8819680</td>\n",
       "      <td>8819752</td>\n",
       "      <td>WBGene00023118</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X</td>\n",
       "      <td>410390</td>\n",
       "      <td>410462</td>\n",
       "      <td>WBGene00023117</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>X</td>\n",
       "      <td>3994601</td>\n",
       "      <td>3994678</td>\n",
       "      <td>WBGene00023116</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>IV</td>\n",
       "      <td>3524818</td>\n",
       "      <td>3524891</td>\n",
       "      <td>WBGene00023115</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>II</td>\n",
       "      <td>4565015</td>\n",
       "      <td>4565087</td>\n",
       "      <td>WBGene00023113</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>II</td>\n",
       "      <td>4565388</td>\n",
       "      <td>4565460</td>\n",
       "      <td>WBGene00023112</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>X</td>\n",
       "      <td>3035975</td>\n",
       "      <td>3036047</td>\n",
       "      <td>WBGene00023111</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>X</td>\n",
       "      <td>3036163</td>\n",
       "      <td>3036235</td>\n",
       "      <td>WBGene00023110</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>X</td>\n",
       "      <td>8640793</td>\n",
       "      <td>8640866</td>\n",
       "      <td>WBGene00022989</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>X</td>\n",
       "      <td>8650708</td>\n",
       "      <td>8650781</td>\n",
       "      <td>WBGene00022988</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>X</td>\n",
       "      <td>8638921</td>\n",
       "      <td>8638993</td>\n",
       "      <td>WBGene00022986</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>I</td>\n",
       "      <td>9320177</td>\n",
       "      <td>9320249</td>\n",
       "      <td>WBGene00012377</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>IV</td>\n",
       "      <td>3552188</td>\n",
       "      <td>3552260</td>\n",
       "      <td>WBGene00022894</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>IV</td>\n",
       "      <td>322631</td>\n",
       "      <td>322703</td>\n",
       "      <td>WBGene00023092</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>III</td>\n",
       "      <td>6513936</td>\n",
       "      <td>6514018</td>\n",
       "      <td>WBGene00023093</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>X</td>\n",
       "      <td>8623997</td>\n",
       "      <td>8624069</td>\n",
       "      <td>WBGene00023090</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>X</td>\n",
       "      <td>8623763</td>\n",
       "      <td>8623835</td>\n",
       "      <td>WBGene00023091</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>V</td>\n",
       "      <td>6559756</td>\n",
       "      <td>6559828</td>\n",
       "      <td>WBGene00023096</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>III</td>\n",
       "      <td>6888805</td>\n",
       "      <td>6888877</td>\n",
       "      <td>WBGene00023097</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>IV</td>\n",
       "      <td>1514591</td>\n",
       "      <td>1514663</td>\n",
       "      <td>WBGene00023094</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>V</td>\n",
       "      <td>6551853</td>\n",
       "      <td>6551935</td>\n",
       "      <td>WBGene00023095</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>III</td>\n",
       "      <td>6888609</td>\n",
       "      <td>6888681</td>\n",
       "      <td>WBGene00023098</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>III</td>\n",
       "      <td>6887727</td>\n",
       "      <td>6887800</td>\n",
       "      <td>WBGene00023099</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>IV</td>\n",
       "      <td>13870264</td>\n",
       "      <td>13870336</td>\n",
       "      <td>WBGene00009206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>IV</td>\n",
       "      <td>5967159</td>\n",
       "      <td>5967232</td>\n",
       "      <td>WBGene00022903</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>X</td>\n",
       "      <td>1589666</td>\n",
       "      <td>1589740</td>\n",
       "      <td>WBGene00022902</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>X</td>\n",
       "      <td>1590079</td>\n",
       "      <td>1590153</td>\n",
       "      <td>WBGene00022901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>IV</td>\n",
       "      <td>15311117</td>\n",
       "      <td>15311199</td>\n",
       "      <td>WBGene00014604</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>IV</td>\n",
       "      <td>15604311</td>\n",
       "      <td>15604383</td>\n",
       "      <td>WBGene00014607</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>606</th>\n",
       "      <td>IV</td>\n",
       "      <td>15793414</td>\n",
       "      <td>15793487</td>\n",
       "      <td>WBGene00014606</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>III</td>\n",
       "      <td>11554824</td>\n",
       "      <td>11554897</td>\n",
       "      <td>WBGene00014600</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>IV</td>\n",
       "      <td>15318477</td>\n",
       "      <td>15318559</td>\n",
       "      <td>WBGene00014603</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>IV</td>\n",
       "      <td>15235387</td>\n",
       "      <td>15235460</td>\n",
       "      <td>WBGene00014602</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>IV</td>\n",
       "      <td>15933352</td>\n",
       "      <td>15933424</td>\n",
       "      <td>WBGene00014609</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>IV</td>\n",
       "      <td>15932009</td>\n",
       "      <td>15932081</td>\n",
       "      <td>WBGene00014608</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>X</td>\n",
       "      <td>14010252</td>\n",
       "      <td>14010324</td>\n",
       "      <td>WBGene00014274</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>X</td>\n",
       "      <td>11806587</td>\n",
       "      <td>11806660</td>\n",
       "      <td>WBGene00014276</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>III</td>\n",
       "      <td>1164799</td>\n",
       "      <td>1164873</td>\n",
       "      <td>WBGene00023197</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>III</td>\n",
       "      <td>1163353</td>\n",
       "      <td>1163427</td>\n",
       "      <td>WBGene00023196</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>III</td>\n",
       "      <td>1424490</td>\n",
       "      <td>1424562</td>\n",
       "      <td>WBGene00023195</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>X</td>\n",
       "      <td>1764253</td>\n",
       "      <td>1764327</td>\n",
       "      <td>WBGene00023194</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>IV</td>\n",
       "      <td>6481160</td>\n",
       "      <td>6481233</td>\n",
       "      <td>WBGene00023192</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>III</td>\n",
       "      <td>2711710</td>\n",
       "      <td>2711782</td>\n",
       "      <td>WBGene00023190</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>I</td>\n",
       "      <td>946043</td>\n",
       "      <td>946115</td>\n",
       "      <td>WBGene00023199</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>I</td>\n",
       "      <td>1575719</td>\n",
       "      <td>1575792</td>\n",
       "      <td>WBGene00023198</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>13274</td>\n",
       "      <td>13327</td>\n",
       "      <td>WBGene00014473</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>9592</td>\n",
       "      <td>9647</td>\n",
       "      <td>WBGene00014470</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>10347</td>\n",
       "      <td>10401</td>\n",
       "      <td>WBGene00014471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>X</td>\n",
       "      <td>13307033</td>\n",
       "      <td>13307106</td>\n",
       "      <td>WBGene00014478</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>X</td>\n",
       "      <td>15200816</td>\n",
       "      <td>15200888</td>\n",
       "      <td>WBGene00014515</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>X</td>\n",
       "      <td>15201006</td>\n",
       "      <td>15201078</td>\n",
       "      <td>WBGene00014514</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>II</td>\n",
       "      <td>9773653</td>\n",
       "      <td>9773728</td>\n",
       "      <td>WBGene00014513</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>I</td>\n",
       "      <td>9327636</td>\n",
       "      <td>9327708</td>\n",
       "      <td>WBGene00012653</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>V</td>\n",
       "      <td>17317878</td>\n",
       "      <td>17317962</td>\n",
       "      <td>WBGene00014362</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>X</td>\n",
       "      <td>15241354</td>\n",
       "      <td>15241427</td>\n",
       "      <td>WBGene00014369</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>I</td>\n",
       "      <td>11505550</td>\n",
       "      <td>11505622</td>\n",
       "      <td>WBGene00014363</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>II</td>\n",
       "      <td>9566240</td>\n",
       "      <td>9566313</td>\n",
       "      <td>WBGene00014365</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>634 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end            name score strand\n",
       "0        X   8409295   8409377  WBGene00022981     .      -\n",
       "1        X   8417911   8417993  WBGene00022980     .      -\n",
       "2        X   8820184   8820256  WBGene00023119     .      +\n",
       "3        X   8819680   8819752  WBGene00023118     .      +\n",
       "4        X    410390    410462  WBGene00023117     .      +\n",
       "5        X   3994601   3994678  WBGene00023116     .      +\n",
       "6       IV   3524818   3524891  WBGene00023115     .      +\n",
       "7       II   4565015   4565087  WBGene00023113     .      -\n",
       "8       II   4565388   4565460  WBGene00023112     .      +\n",
       "9        X   3035975   3036047  WBGene00023111     .      -\n",
       "10       X   3036163   3036235  WBGene00023110     .      +\n",
       "11       X   8640793   8640866  WBGene00022989     .      -\n",
       "12       X   8650708   8650781  WBGene00022988     .      -\n",
       "13       X   8638921   8638993  WBGene00022986     .      +\n",
       "14       I   9320177   9320249  WBGene00012377     .      +\n",
       "15      IV   3552188   3552260  WBGene00022894     .      -\n",
       "16      IV    322631    322703  WBGene00023092     .      -\n",
       "17     III   6513936   6514018  WBGene00023093     .      -\n",
       "18       X   8623997   8624069  WBGene00023090     .      -\n",
       "19       X   8623763   8623835  WBGene00023091     .      -\n",
       "20       V   6559756   6559828  WBGene00023096     .      +\n",
       "21     III   6888805   6888877  WBGene00023097     .      +\n",
       "22      IV   1514591   1514663  WBGene00023094     .      +\n",
       "23       V   6551853   6551935  WBGene00023095     .      -\n",
       "24     III   6888609   6888681  WBGene00023098     .      -\n",
       "25     III   6887727   6887800  WBGene00023099     .      -\n",
       "26      IV  13870264  13870336  WBGene00009206     .      +\n",
       "27      IV   5967159   5967232  WBGene00022903     .      +\n",
       "28       X   1589666   1589740  WBGene00022902     .      -\n",
       "29       X   1590079   1590153  WBGene00022901     .      -\n",
       "..     ...       ...       ...             ...   ...    ...\n",
       "604     IV  15311117  15311199  WBGene00014604     .      -\n",
       "605     IV  15604311  15604383  WBGene00014607     .      -\n",
       "606     IV  15793414  15793487  WBGene00014606     .      +\n",
       "607    III  11554824  11554897  WBGene00014600     .      +\n",
       "608     IV  15318477  15318559  WBGene00014603     .      -\n",
       "609     IV  15235387  15235460  WBGene00014602     .      +\n",
       "610     IV  15933352  15933424  WBGene00014609     .      +\n",
       "611     IV  15932009  15932081  WBGene00014608     .      +\n",
       "612      X  14010252  14010324  WBGene00014274     .      +\n",
       "613      X  11806587  11806660  WBGene00014276     .      -\n",
       "614    III   1164799   1164873  WBGene00023197     .      -\n",
       "615    III   1163353   1163427  WBGene00023196     .      -\n",
       "616    III   1424490   1424562  WBGene00023195     .      +\n",
       "617      X   1764253   1764327  WBGene00023194     .      -\n",
       "618     IV   6481160   6481233  WBGene00023192     .      +\n",
       "619    III   2711710   2711782  WBGene00023190     .      +\n",
       "620      I    946043    946115  WBGene00023199     .      -\n",
       "621      I   1575719   1575792  WBGene00023198     .      +\n",
       "622  MtDNA     13274     13327  WBGene00014473     .      +\n",
       "623  MtDNA      9592      9647  WBGene00014470     .      +\n",
       "624  MtDNA     10347     10401  WBGene00014471     .      +\n",
       "625      X  13307033  13307106  WBGene00014478     .      +\n",
       "626      X  15200816  15200888  WBGene00014515     .      -\n",
       "627      X  15201006  15201078  WBGene00014514     .      +\n",
       "628     II   9773653   9773728  WBGene00014513     .      +\n",
       "629      I   9327636   9327708  WBGene00012653     .      +\n",
       "630      V  17317878  17317962  WBGene00014362     .      -\n",
       "631      X  15241354  15241427  WBGene00014369     .      -\n",
       "632      I  11505550  11505622  WBGene00014363     .      +\n",
       "633     II   9566240   9566313  WBGene00014365     .      +\n",
       "\n",
       "[634 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/WBcel235/annotation/Caenorhabditis_elegans.WBcel235.91.gffutils.rRNA_sites.bed)>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites = []\n",
    "rRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'rRNA' in transcript.attributes['gene_biotype']:\n",
    "            rRNA_sites.append(transcript)\n",
    "    #renamed_rRNA_sites = rename_regions(rRNA_sites, gene_id)\n",
    "    rRNA_bed += create_bed(rRNA_sites)\n",
    "rRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(rRNA_bed.split('\\n'))))\n",
    "\n",
    "rRNA_sites_bedtool = pybedtools.BedTool(rRNA_bed, from_string=True)\n",
    "rRNA_sites_bedtool.remove_invalid().sort().saveas('{}.rRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V</td>\n",
       "      <td>17120967</td>\n",
       "      <td>17121086</td>\n",
       "      <td>WBGene00014621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>897</td>\n",
       "      <td>1593</td>\n",
       "      <td>WBGene00014454</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>V</td>\n",
       "      <td>17120012</td>\n",
       "      <td>17120111</td>\n",
       "      <td>WBGene00077468</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>V</td>\n",
       "      <td>17121943</td>\n",
       "      <td>17122062</td>\n",
       "      <td>WBGene00077469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>V</td>\n",
       "      <td>17117862</td>\n",
       "      <td>17117981</td>\n",
       "      <td>WBGene00077466</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>V</td>\n",
       "      <td>17118836</td>\n",
       "      <td>17118935</td>\n",
       "      <td>WBGene00077467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>V</td>\n",
       "      <td>17115902</td>\n",
       "      <td>17116021</td>\n",
       "      <td>WBGene00077465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>I</td>\n",
       "      <td>15064837</td>\n",
       "      <td>15068346</td>\n",
       "      <td>WBGene00004622</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>V</td>\n",
       "      <td>17428878</td>\n",
       "      <td>17428996</td>\n",
       "      <td>WBGene00189966</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>15064300</td>\n",
       "      <td>15064453</td>\n",
       "      <td>WBGene00004567</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>I</td>\n",
       "      <td>15062082</td>\n",
       "      <td>15063836</td>\n",
       "      <td>WBGene00004512</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>I</td>\n",
       "      <td>15069279</td>\n",
       "      <td>15071033</td>\n",
       "      <td>WBGene00004513</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>6169083</td>\n",
       "      <td>6169195</td>\n",
       "      <td>WBGene00235197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>V</td>\n",
       "      <td>17129375</td>\n",
       "      <td>17129494</td>\n",
       "      <td>WBGene00077475</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>V</td>\n",
       "      <td>17128403</td>\n",
       "      <td>17128522</td>\n",
       "      <td>WBGene00077474</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>V</td>\n",
       "      <td>17131337</td>\n",
       "      <td>17131456</td>\n",
       "      <td>WBGene00077477</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V</td>\n",
       "      <td>17130361</td>\n",
       "      <td>17130480</td>\n",
       "      <td>WBGene00077476</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>V</td>\n",
       "      <td>17123895</td>\n",
       "      <td>17124014</td>\n",
       "      <td>WBGene00077471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>V</td>\n",
       "      <td>17122919</td>\n",
       "      <td>17123038</td>\n",
       "      <td>WBGene00077470</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>V</td>\n",
       "      <td>17125847</td>\n",
       "      <td>17125966</td>\n",
       "      <td>WBGene00077473</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>V</td>\n",
       "      <td>17124871</td>\n",
       "      <td>17124990</td>\n",
       "      <td>WBGene00077472</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MtDNA</td>\n",
       "      <td>10402</td>\n",
       "      <td>11354</td>\n",
       "      <td>WBGene00014472</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end            name score strand\n",
       "0       V  17120967  17121086  WBGene00014621     .      +\n",
       "1   MtDNA       897      1593  WBGene00014454     .      +\n",
       "2       V  17120012  17120111  WBGene00077468     .      +\n",
       "3       V  17121943  17122062  WBGene00077469     .      +\n",
       "4       V  17117862  17117981  WBGene00077466     .      +\n",
       "5       V  17118836  17118935  WBGene00077467     .      +\n",
       "6       V  17115902  17116021  WBGene00077465     .      +\n",
       "7       I  15064837  15068346  WBGene00004622     .      +\n",
       "8       V  17428878  17428996  WBGene00189966     .      +\n",
       "9       I  15064300  15064453  WBGene00004567     .      +\n",
       "10      I  15062082  15063836  WBGene00004512     .      +\n",
       "11      I  15069279  15071033  WBGene00004513     .      +\n",
       "12      I   6169083   6169195  WBGene00235197     .      +\n",
       "13      V  17129375  17129494  WBGene00077475     .      +\n",
       "14      V  17128403  17128522  WBGene00077474     .      +\n",
       "15      V  17131337  17131456  WBGene00077477     .      +\n",
       "16      V  17130361  17130480  WBGene00077476     .      +\n",
       "17      V  17123895  17124014  WBGene00077471     .      +\n",
       "18      V  17122919  17123038  WBGene00077470     .      +\n",
       "19      V  17125847  17125966  WBGene00077473     .      +\n",
       "20      V  17124871  17124990  WBGene00077472     .      +\n",
       "21  MtDNA     10402     11354  WBGene00014472     .      +"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites_bedtool. to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
