{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:08:09.166605Z",
     "start_time": "2017-09-28T03:08:07.636901Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict, OrderedDict\n",
    "import warnings\n",
    "import logging\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import copy\n",
    "import re\n",
    "from gffutils.pybedtools_integration import tsses\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:11:54.277345Z",
     "start_time": "2017-09-28T03:11:54.265908Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gtf'\n",
    "gtf_db = '/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gtf.db'\n",
    "prefix = '/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils'\n",
    "chrsizes = '/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/fasta/Gallus_gallus.Gallus_gallus-5.0.91.dna.toplevel.sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:15:43.298837Z",
     "start_time": "2017-09-28T03:11:55.256580Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.create_db(gtf, dbfn=gtf_db, disable_infer_genes=True, disable_infer_transcripts=True, merge_strategy='merge', force=True)\n",
    "def create_gene_dict(db):\n",
    "    '''\n",
    "    Store each feature line db.all_features() as a dict of dicts\n",
    "    '''\n",
    "    gene_dict = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))\n",
    "    for line_no, feature in enumerate(db.all_features()):\n",
    "        gene_ids = feature.attributes['gene_id']\n",
    "        feature_type = feature.featuretype\n",
    "        if feature_type == 'gene':\n",
    "            if len(gene_ids)!=1:\n",
    "                logging.warning('Found multiple gene_ids on line {} in gtf'.format(line_no))\n",
    "                break\n",
    "            else:\n",
    "                gene_id = gene_ids[0]\n",
    "                gene_dict[gene_id]['gene'] = feature\n",
    "        else:\n",
    "            transcript_ids = feature.attributes['transcript_id']\n",
    "\n",
    "            for gene_id in gene_ids:\n",
    "                for transcript_id in transcript_ids:\n",
    "                    gene_dict[gene_id][transcript_id][feature_type].append(feature)\n",
    "    return gene_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.280784Z",
     "start_time": "2017-09-28T03:15:43.304709Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(gtf_db, keep_order=True)\n",
    "gene_dict = create_gene_dict(db)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.353531Z",
     "start_time": "2017-09-28T03:16:43.285946Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS\n",
      "exon\n",
      "five_prime_utr\n",
      "gene\n",
      "start_codon\n",
      "stop_codon\n",
      "three_prime_utr\n",
      "transcript\n"
     ]
    }
   ],
   "source": [
    "for x in db.featuretypes():\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.461123Z",
     "start_time": "2017-09-28T03:16:43.357554Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "def get_gene_list(gene_dict):\n",
    "    return list(set(gene_dict.keys()))\n",
    "\n",
    "def get_UTR_regions(gene_dict, gene_id, transcript, cds):\n",
    "    if len(cds)==0:\n",
    "        return [], []\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    utrs = gene_dict[gene_id][transcript]['UTR']\n",
    "    first_cds = cds[0]\n",
    "    last_cds = cds[-1]\n",
    "    for utr in utrs:\n",
    "        ## Push all cds at once\n",
    "        ## Sort later to remove duplicates\n",
    "        strand = utr.strand\n",
    "        if strand == '+':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')\n",
    "        elif strand == '-':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)                \n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')    \n",
    "    return utr5_regions, utr3_regions\n",
    "    \n",
    "def create_bed(regions, bedtype='0'):\n",
    "    '''Create bed from list of regions\n",
    "    bedtype: 0 or 1\n",
    "        0-Based or 1-based coordinate of the BED\n",
    "    '''\n",
    "    bedstr = ''\n",
    "    for region in regions:\n",
    "        assert len(region.attributes['gene_id']) == 1\n",
    "        ## GTF start is 1-based, so shift by one while writing \n",
    "        ## to 0-based BED format\n",
    "        if bedtype == '0':\n",
    "            start = region.start - 1\n",
    "        else:\n",
    "            start = region.start\n",
    "        bedstr += '{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(region.chrom,\n",
    "                                             start,\n",
    "                                             region.stop,\n",
    "                                             re.sub('\\.\\d+', '', region.attributes['gene_id'][0]),\n",
    "                                             '.',\n",
    "                                             region.strand)\n",
    "    # Remove duplicates\n",
    "    bedstr = '\\n'.join(list(OrderedDict.fromkeys(bedstr.split('\\n'))))\n",
    "    return bedstr\n",
    "\n",
    "def rename_regions(regions, gene_id):\n",
    "    regions = list(regions)\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    for region in regions:\n",
    "        region.attributes['gene_id'] = gene_id\n",
    "    return regions\n",
    "\n",
    "def merge_regions(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start))\n",
    "    return merged\n",
    "\n",
    "def merge_regions_nostrand(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start), ignore_strand=True)\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:17:38.562657Z",
     "start_time": "2017-09-28T03:16:43.463672Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.cds.bed)>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_bed = ''\n",
    "utr3_bed = ''\n",
    "\n",
    "\n",
    "\n",
    "gene_bed = ''\n",
    "exon_bed = ''\n",
    "intron_bed = ''\n",
    "start_codon_bed = ''\n",
    "stop_codon_bed = ''\n",
    "cds_bed = ''\n",
    "\n",
    "gene_list = []\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    gene_list.append(gene_dict[gene_id]['gene'])\n",
    "    \n",
    "    utr5_regions, utr3_regions = [], []\n",
    "    exon_regions, intron_regions = [], []\n",
    "    star_codon_regions, stop_codon_regions = [], []\n",
    "    cds_regions = []\n",
    "    \n",
    "    for feature in gene_dict[gene_id].keys():\n",
    "        if feature == 'gene':\n",
    "            continue\n",
    "        cds = list(gene_dict[gene_id][feature]['CDS'])\n",
    "        exons = list(gene_dict[gene_id][feature]['exon'])\n",
    "        merged_exons = merge_regions(db, exons)\n",
    "        introns = db.interfeatures(merged_exons)\n",
    "        utr5_region = list(gene_dict[gene_id][feature]['five_prime_utr'])\n",
    "        utr3_region = list(gene_dict[gene_id][feature]['three_prime_utr'])\n",
    "        utr5_regions += utr5_region\n",
    "        utr3_regions += utr3_region\n",
    "        exon_regions += exons\n",
    "        intron_regions += introns\n",
    "        cds_regions += cds\n",
    "        \n",
    "    merged_utr5 = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5 = rename_regions(merged_utr5, gene_id)\n",
    "    \n",
    "    merged_utr3 = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3 = rename_regions(merged_utr3, gene_id)\n",
    "    \n",
    "    merged_exons = merge_regions(db, exon_regions)\n",
    "    renamed_exons = rename_regions(merged_exons, gene_id)\n",
    "    \n",
    "    merged_introns = merge_regions(db, intron_regions)\n",
    "    renamed_introns = rename_regions(merged_introns, gene_id)\n",
    "    \n",
    "    merged_cds = merge_regions(db, cds_regions)\n",
    "    renamed_cds = rename_regions(merged_cds, gene_id)\n",
    "    \n",
    "    utr3_bed += create_bed(renamed_utr3)\n",
    "    utr5_bed += create_bed(renamed_utr5)\n",
    "    exon_bed += create_bed(renamed_exons)\n",
    "    intron_bed += create_bed(renamed_introns)\n",
    "    cds_bed += create_bed(renamed_cds)\n",
    "    \n",
    "    \n",
    "gene_bed = create_bed(gene_list)\n",
    "gene_bedtool = pybedtools.BedTool(gene_bed, from_string=True)\n",
    "utr5_bedtool = pybedtools.BedTool(utr5_bed, from_string=True)\n",
    "utr3_bedtool = pybedtools.BedTool(utr3_bed, from_string=True)\n",
    "exon_bedtool = pybedtools.BedTool(exon_bed, from_string=True)\n",
    "intron_bedtool = pybedtools.BedTool(intron_bed, from_string=True)\n",
    "cds_bedtool = pybedtools.BedTool(cds_bed, from_string=True)\n",
    "\n",
    "gene_bedtool.remove_invalid().sort().saveas('{}.genes.bed'.format(prefix))\n",
    "utr5_bedtool.remove_invalid().sort().saveas('{}.UTR5.bed'.format(prefix))\n",
    "utr3_bedtool.remove_invalid().sort().saveas('{}.UTR3.bed'.format(prefix))\n",
    "exon_bedtool.remove_invalid().sort().saveas('{}.exon.bed'.format(prefix))\n",
    "intron_bedtool.remove_invalid().sort().saveas('{}.intron.bed'.format(prefix))\n",
    "cds_bedtool.remove_invalid().sort().saveas('{}.cds.bed'.format(prefix))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T04:25:18.715414Z",
     "start_time": "2017-09-28T04:20:10.158492Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "utr5_region_bed = ''\n",
    "utr3_region_bed = ''\n",
    "\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    for utr5_region in db.children(gene_id, featuretype='five_prime_utr'):\n",
    "        utr5_regions.append(utr5_region)\n",
    "    for utr3_region in db.children(gene_id, featuretype='three_prime_utr'):\n",
    "        utr3_regions.append(utr3_region)\n",
    "    merged_utr5_regions = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5_regions = rename_regions(merged_utr5_regions, gene_id)\n",
    "    merged_utr3_regions = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3_regions = rename_regions(merged_utr3_regions, gene_id)\n",
    "    \n",
    "    utr5_region_bed += create_bed(renamed_utr5_regions)    \n",
    "    utr3_region_bed += create_bed(renamed_utr3_regions)\n",
    "\n",
    "\n",
    "    \n",
    "utr5_region_bedtool = pybedtools.BedTool(utr5_region_bed, from_string=True)\n",
    "utr3_region_bedtool = pybedtools.BedTool(utr3_region_bed, from_string=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:08:46.134982Z",
     "start_time": "2017-09-28T05:08:45.236263Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr3_region.bed)>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_region_bedtool.remove_invalid().sort().saveas('{}.utr5_region.bed'.format(prefix))\n",
    "utr3_region_bedtool.remove_invalid().sort().saveas('{}.utr3_region.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:09:02.138861Z",
     "start_time": "2017-09-28T05:09:02.130605Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr5_region.bed'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'{}.utr5_region.bed'.format(prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:09:21.627640Z",
     "start_time": "2017-09-28T05:09:21.075155Z"
    }
   },
   "outputs": [],
   "source": [
    "!bedSort /home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr3_region.bed /home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr3_region.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!bedSort /home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr5_region.bed /home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.utr5_region.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.tRNA_sites.bed)>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites = []\n",
    "tRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'tRNA' in transcript.attributes['gene_biotype'] or 'Mt_tRNA' in transcript.attributes['transcript_biotype']:\n",
    "            tRNA_sites.append(transcript)\n",
    "    #merged_tRNA_sites = merge_regions_nostrand(db, tRNA_sites)\n",
    "    #renamed_tRNA_sites = rename_regions(merged_tRNA_sites, gene_id)\n",
    "    tRNA_bed += create_bed(tRNA_sites)\n",
    "\n",
    "tRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(tRNA_bed.split('\\n'))))\n",
    "tRNA_sites_bedtool = pybedtools.BedTool(tRNA_bed, from_string=True)\n",
    "tRNA_sites_bedtool.remove_invalid().sort().saveas('{}.tRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>5101</td>\n",
       "      <td>5172</td>\n",
       "      <td>ENSGALG00000029193</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>9015</td>\n",
       "      <td>9083</td>\n",
       "      <td>ENSGALG00000038283</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>8260</td>\n",
       "      <td>8329</td>\n",
       "      <td>ENSGALG00000031197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>6433</td>\n",
       "      <td>6506</td>\n",
       "      <td>ENSGALG00000033139</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>12863</td>\n",
       "      <td>12932</td>\n",
       "      <td>ENSGALG00000042903</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MT</td>\n",
       "      <td>12999</td>\n",
       "      <td>13070</td>\n",
       "      <td>ENSGALG00000036970</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MT</td>\n",
       "      <td>12933</td>\n",
       "      <td>12998</td>\n",
       "      <td>ENSGALG00000034022</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MT</td>\n",
       "      <td>6507</td>\n",
       "      <td>6573</td>\n",
       "      <td>ENSGALG00000034813</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MT</td>\n",
       "      <td>6361</td>\n",
       "      <td>6430</td>\n",
       "      <td>ENSGALG00000035392</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MT</td>\n",
       "      <td>16707</td>\n",
       "      <td>16775</td>\n",
       "      <td>ENSGALG00000039249</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MT</td>\n",
       "      <td>3966</td>\n",
       "      <td>4040</td>\n",
       "      <td>ENSGALG00000040296</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MT</td>\n",
       "      <td>8123</td>\n",
       "      <td>8258</td>\n",
       "      <td>ENSGALG00000033462</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MT</td>\n",
       "      <td>5024</td>\n",
       "      <td>5096</td>\n",
       "      <td>ENSGALG00000035975</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MT</td>\n",
       "      <td>5171</td>\n",
       "      <td>5240</td>\n",
       "      <td>ENSGALG00000038760</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MT</td>\n",
       "      <td>2272</td>\n",
       "      <td>2345</td>\n",
       "      <td>ENSGALG00000032059</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MT</td>\n",
       "      <td>6279</td>\n",
       "      <td>6355</td>\n",
       "      <td>ENSGALG00000035685</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MT</td>\n",
       "      <td>10707</td>\n",
       "      <td>10775</td>\n",
       "      <td>ENSGALG00000038243</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>MT</td>\n",
       "      <td>1227</td>\n",
       "      <td>1296</td>\n",
       "      <td>ENSGALG00000041922</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MT</td>\n",
       "      <td>11127</td>\n",
       "      <td>11195</td>\n",
       "      <td>ENSGALG00000037369</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MT</td>\n",
       "      <td>16038</td>\n",
       "      <td>16107</td>\n",
       "      <td>ENSGALG00000032370</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MT</td>\n",
       "      <td>6572</td>\n",
       "      <td>6643</td>\n",
       "      <td>ENSGALG00000037641</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MT</td>\n",
       "      <td>16107</td>\n",
       "      <td>16177</td>\n",
       "      <td>ENSGALG00000042677</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom  start    end                name score strand\n",
       "0     MT   5101   5172  ENSGALG00000029193     .      -\n",
       "1     MT   9015   9083  ENSGALG00000038283     .      +\n",
       "2     MT   8260   8329  ENSGALG00000031197     .      +\n",
       "3     MT   6433   6506  ENSGALG00000033139     .      -\n",
       "4     MT  12863  12932  ENSGALG00000042903     .      +\n",
       "5     MT  12999  13070  ENSGALG00000036970     .      +\n",
       "6     MT  12933  12998  ENSGALG00000034022     .      +\n",
       "7     MT   6507   6573  ENSGALG00000034813     .      -\n",
       "8     MT   6361   6430  ENSGALG00000035392     .      -\n",
       "9     MT  16707  16775  ENSGALG00000039249     .      -\n",
       "10    MT   3966   4040  ENSGALG00000040296     .      +\n",
       "11    MT   8123   8258  ENSGALG00000033462     .      -\n",
       "12    MT   5024   5096  ENSGALG00000035975     .      +\n",
       "13    MT   5171   5240  ENSGALG00000038760     .      +\n",
       "14    MT   2272   2345  ENSGALG00000032059     .      +\n",
       "15    MT   6279   6355  ENSGALG00000035685     .      +\n",
       "16    MT  10707  10775  ENSGALG00000038243     .      +\n",
       "17    MT   1227   1296  ENSGALG00000041922     .      +\n",
       "18    MT  11127  11195  ENSGALG00000037369     .      +\n",
       "19    MT  16038  16107  ENSGALG00000032370     .      +\n",
       "20    MT   6572   6643  ENSGALG00000037641     .      -\n",
       "21    MT  16107  16177  ENSGALG00000042677     .      -"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.rRNA_sites.bed)>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites = []\n",
    "rRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'rRNA' in transcript.attributes['gene_biotype']:\n",
    "            rRNA_sites.append(transcript)\n",
    "    #renamed_rRNA_sites = rename_regions(rRNA_sites, gene_id)\n",
    "    rRNA_bed += create_bed(rRNA_sites)\n",
    "rRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(rRNA_bed.split('\\n'))))\n",
    "rRNA_sites_bedtool = pybedtools.BedTool(rRNA_bed, from_string=True)\n",
    "rRNA_sites_bedtool.remove_invalid().sort().saveas('{}.rRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AADN04016707.1</td>\n",
       "      <td>564</td>\n",
       "      <td>682</td>\n",
       "      <td>ENSGALG00000031236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AADN04003868.1</td>\n",
       "      <td>7975</td>\n",
       "      <td>8128</td>\n",
       "      <td>ENSGALG00000032920</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AADN04001404.1</td>\n",
       "      <td>14906</td>\n",
       "      <td>15029</td>\n",
       "      <td>ENSGALG00000039272</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>28184027</td>\n",
       "      <td>28184138</td>\n",
       "      <td>ENSGALG00000027900</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AADN04001404.1</td>\n",
       "      <td>2298</td>\n",
       "      <td>2417</td>\n",
       "      <td>ENSGALG00000036591</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AADN04007760.1</td>\n",
       "      <td>9983</td>\n",
       "      <td>10135</td>\n",
       "      <td>ENSGALG00000037964</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AADN04017093.1</td>\n",
       "      <td>2714</td>\n",
       "      <td>2866</td>\n",
       "      <td>ENSGALG00000037963</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AADN04003312.1</td>\n",
       "      <td>248</td>\n",
       "      <td>367</td>\n",
       "      <td>ENSGALG00000032451</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AADN04001404.1</td>\n",
       "      <td>25686</td>\n",
       "      <td>25804</td>\n",
       "      <td>ENSGALG00000039768</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AADN04003312.1</td>\n",
       "      <td>12866</td>\n",
       "      <td>12980</td>\n",
       "      <td>ENSGALG00000029463</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AADN04002199.1</td>\n",
       "      <td>11011</td>\n",
       "      <td>11130</td>\n",
       "      <td>ENSGALG00000035984</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AADN04003863.1</td>\n",
       "      <td>8576</td>\n",
       "      <td>8729</td>\n",
       "      <td>ENSGALG00000033913</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AADN04006969.1</td>\n",
       "      <td>8399</td>\n",
       "      <td>8552</td>\n",
       "      <td>ENSGALG00000036940</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AADN04004372.1</td>\n",
       "      <td>9347</td>\n",
       "      <td>9499</td>\n",
       "      <td>ENSGALG00000029838</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>166149900</td>\n",
       "      <td>166150027</td>\n",
       "      <td>ENSGALG00000017789</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AADN04004449.1</td>\n",
       "      <td>5267</td>\n",
       "      <td>5386</td>\n",
       "      <td>ENSGALG00000030448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AADN04005700.1</td>\n",
       "      <td>3890</td>\n",
       "      <td>4045</td>\n",
       "      <td>ENSGALG00000029977</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1</td>\n",
       "      <td>45543827</td>\n",
       "      <td>45543933</td>\n",
       "      <td>ENSGALG00000028788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AADN04002018.1</td>\n",
       "      <td>7780</td>\n",
       "      <td>7899</td>\n",
       "      <td>ENSGALG00000029786</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>9</td>\n",
       "      <td>937970</td>\n",
       "      <td>938092</td>\n",
       "      <td>ENSGALG00000039601</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AADN04005104.1</td>\n",
       "      <td>12382</td>\n",
       "      <td>12535</td>\n",
       "      <td>ENSGALG00000041531</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KQ759483.1</td>\n",
       "      <td>693765</td>\n",
       "      <td>693883</td>\n",
       "      <td>ENSGALG00000038956</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AADN04005797.1</td>\n",
       "      <td>10116</td>\n",
       "      <td>10269</td>\n",
       "      <td>ENSGALG00000029549</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AADN04002199.1</td>\n",
       "      <td>19365</td>\n",
       "      <td>19484</td>\n",
       "      <td>ENSGALG00000037013</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AADN04004449.1</td>\n",
       "      <td>11791</td>\n",
       "      <td>11910</td>\n",
       "      <td>ENSGALG00000038288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AADN04002018.1</td>\n",
       "      <td>12150</td>\n",
       "      <td>12269</td>\n",
       "      <td>ENSGALG00000038105</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AADN04000891.1</td>\n",
       "      <td>18394</td>\n",
       "      <td>18513</td>\n",
       "      <td>ENSGALG00000032076</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AADN04002199.1</td>\n",
       "      <td>163</td>\n",
       "      <td>282</td>\n",
       "      <td>ENSGALG00000037931</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>AADN04003737.1</td>\n",
       "      <td>9352</td>\n",
       "      <td>9505</td>\n",
       "      <td>ENSGALG00000041381</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AADN04003187.1</td>\n",
       "      <td>10189</td>\n",
       "      <td>10342</td>\n",
       "      <td>ENSGALG00000040262</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>AADN04002587.1</td>\n",
       "      <td>11093</td>\n",
       "      <td>11230</td>\n",
       "      <td>ENSGALG00000039667</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>AADN04001089.1</td>\n",
       "      <td>1384</td>\n",
       "      <td>1536</td>\n",
       "      <td>ENSGALG00000029561</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>9</td>\n",
       "      <td>935922</td>\n",
       "      <td>936041</td>\n",
       "      <td>ENSGALG00000037069</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>8</td>\n",
       "      <td>19457840</td>\n",
       "      <td>19457958</td>\n",
       "      <td>ENSGALG00000027857</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>AADN04001404.1</td>\n",
       "      <td>4389</td>\n",
       "      <td>4508</td>\n",
       "      <td>ENSGALG00000043549</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>AADN04001305.1</td>\n",
       "      <td>3010</td>\n",
       "      <td>3162</td>\n",
       "      <td>ENSGALG00000043542</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>AADN04006451.1</td>\n",
       "      <td>7165</td>\n",
       "      <td>7318</td>\n",
       "      <td>ENSGALG00000029496</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>AADN04002018.1</td>\n",
       "      <td>9965</td>\n",
       "      <td>10084</td>\n",
       "      <td>ENSGALG00000031750</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>AADN04005188.1</td>\n",
       "      <td>2081</td>\n",
       "      <td>2199</td>\n",
       "      <td>ENSGALG00000038067</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>AADN04006374.1</td>\n",
       "      <td>11848</td>\n",
       "      <td>11982</td>\n",
       "      <td>ENSGALG00000037550</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>2</td>\n",
       "      <td>135524230</td>\n",
       "      <td>135524346</td>\n",
       "      <td>ENSGALG00000025601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>AADN04011704.1</td>\n",
       "      <td>3305</td>\n",
       "      <td>3425</td>\n",
       "      <td>ENSGALG00000041250</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>9</td>\n",
       "      <td>944328</td>\n",
       "      <td>944447</td>\n",
       "      <td>ENSGALG00000029609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>AADN04000891.1</td>\n",
       "      <td>20578</td>\n",
       "      <td>20697</td>\n",
       "      <td>ENSGALG00000036818</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>AADN04003312.1</td>\n",
       "      <td>14520</td>\n",
       "      <td>14639</td>\n",
       "      <td>ENSGALG00000032018</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>AADN04009122.1</td>\n",
       "      <td>8449</td>\n",
       "      <td>8601</td>\n",
       "      <td>ENSGALG00000031378</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>AADN04004293.1</td>\n",
       "      <td>6937</td>\n",
       "      <td>7089</td>\n",
       "      <td>ENSGALG00000042701</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>AADN04011704.1</td>\n",
       "      <td>7432</td>\n",
       "      <td>7552</td>\n",
       "      <td>ENSGALG00000029869</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>AADN04002836.1</td>\n",
       "      <td>1555</td>\n",
       "      <td>1708</td>\n",
       "      <td>ENSGALG00000040165</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>AADN04000891.1</td>\n",
       "      <td>22744</td>\n",
       "      <td>22863</td>\n",
       "      <td>ENSGALG00000029288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>AADN04005188.1</td>\n",
       "      <td>12747</td>\n",
       "      <td>12866</td>\n",
       "      <td>ENSGALG00000041326</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>AADN04004449.1</td>\n",
       "      <td>9613</td>\n",
       "      <td>9732</td>\n",
       "      <td>ENSGALG00000043663</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>AADN04009588.1</td>\n",
       "      <td>5404</td>\n",
       "      <td>5522</td>\n",
       "      <td>ENSGALG00000043431</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>AADN04004449.1</td>\n",
       "      <td>1369</td>\n",
       "      <td>1488</td>\n",
       "      <td>ENSGALG00000040257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>AADN04003913.1</td>\n",
       "      <td>12476</td>\n",
       "      <td>12629</td>\n",
       "      <td>ENSGALG00000038196</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>AADN04001404.1</td>\n",
       "      <td>23483</td>\n",
       "      <td>23602</td>\n",
       "      <td>ENSGALG00000031897</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>AADN04003312.1</td>\n",
       "      <td>8596</td>\n",
       "      <td>8715</td>\n",
       "      <td>ENSGALG00000030604</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>AADN04003312.1</td>\n",
       "      <td>2313</td>\n",
       "      <td>2366</td>\n",
       "      <td>ENSGALG00000031407</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>AADN04011388.1</td>\n",
       "      <td>5207</td>\n",
       "      <td>5360</td>\n",
       "      <td>ENSGALG00000035877</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>AADN04007387.1</td>\n",
       "      <td>7607</td>\n",
       "      <td>7756</td>\n",
       "      <td>ENSGALG00000036379</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              chrom      start        end                name score strand\n",
       "0    AADN04016707.1        564        682  ENSGALG00000031236     .      -\n",
       "1    AADN04003868.1       7975       8128  ENSGALG00000032920     .      +\n",
       "2    AADN04001404.1      14906      15029  ENSGALG00000039272     .      +\n",
       "3                 2   28184027   28184138  ENSGALG00000027900     .      -\n",
       "4    AADN04001404.1       2298       2417  ENSGALG00000036591     .      +\n",
       "5    AADN04007760.1       9983      10135  ENSGALG00000037964     .      -\n",
       "6    AADN04017093.1       2714       2866  ENSGALG00000037963     .      +\n",
       "7    AADN04003312.1        248        367  ENSGALG00000032451     .      +\n",
       "8    AADN04001404.1      25686      25804  ENSGALG00000039768     .      +\n",
       "9    AADN04003312.1      12866      12980  ENSGALG00000029463     .      +\n",
       "10   AADN04002199.1      11011      11130  ENSGALG00000035984     .      +\n",
       "11   AADN04003863.1       8576       8729  ENSGALG00000033913     .      +\n",
       "12   AADN04006969.1       8399       8552  ENSGALG00000036940     .      -\n",
       "13   AADN04004372.1       9347       9499  ENSGALG00000029838     .      +\n",
       "14                1  166149900  166150027  ENSGALG00000017789     .      -\n",
       "15   AADN04004449.1       5267       5386  ENSGALG00000030448     .      +\n",
       "16   AADN04005700.1       3890       4045  ENSGALG00000029977     .      +\n",
       "17                1   45543827   45543933  ENSGALG00000028788     .      -\n",
       "18   AADN04002018.1       7780       7899  ENSGALG00000029786     .      +\n",
       "19                9     937970     938092  ENSGALG00000039601     .      -\n",
       "20   AADN04005104.1      12382      12535  ENSGALG00000041531     .      -\n",
       "21       KQ759483.1     693765     693883  ENSGALG00000038956     .      -\n",
       "22   AADN04005797.1      10116      10269  ENSGALG00000029549     .      -\n",
       "23   AADN04002199.1      19365      19484  ENSGALG00000037013     .      +\n",
       "24   AADN04004449.1      11791      11910  ENSGALG00000038288     .      +\n",
       "25   AADN04002018.1      12150      12269  ENSGALG00000038105     .      +\n",
       "26   AADN04000891.1      18394      18513  ENSGALG00000032076     .      +\n",
       "27   AADN04002199.1        163        282  ENSGALG00000037931     .      +\n",
       "28   AADN04003737.1       9352       9505  ENSGALG00000041381     .      -\n",
       "29   AADN04003187.1      10189      10342  ENSGALG00000040262     .      -\n",
       "..              ...        ...        ...                 ...   ...    ...\n",
       "173  AADN04002587.1      11093      11230  ENSGALG00000039667     .      +\n",
       "174  AADN04001089.1       1384       1536  ENSGALG00000029561     .      +\n",
       "175               9     935922     936041  ENSGALG00000037069     .      -\n",
       "176               8   19457840   19457958  ENSGALG00000027857     .      -\n",
       "177  AADN04001404.1       4389       4508  ENSGALG00000043549     .      +\n",
       "178  AADN04001305.1       3010       3162  ENSGALG00000043542     .      -\n",
       "179  AADN04006451.1       7165       7318  ENSGALG00000029496     .      +\n",
       "180  AADN04002018.1       9965      10084  ENSGALG00000031750     .      +\n",
       "181  AADN04005188.1       2081       2199  ENSGALG00000038067     .      +\n",
       "182  AADN04006374.1      11848      11982  ENSGALG00000037550     .      -\n",
       "183               2  135524230  135524346  ENSGALG00000025601     .      +\n",
       "184  AADN04011704.1       3305       3425  ENSGALG00000041250     .      +\n",
       "185               9     944328     944447  ENSGALG00000029609     .      -\n",
       "186  AADN04000891.1      20578      20697  ENSGALG00000036818     .      +\n",
       "187  AADN04003312.1      14520      14639  ENSGALG00000032018     .      +\n",
       "188  AADN04009122.1       8449       8601  ENSGALG00000031378     .      -\n",
       "189  AADN04004293.1       6937       7089  ENSGALG00000042701     .      +\n",
       "190  AADN04011704.1       7432       7552  ENSGALG00000029869     .      +\n",
       "191  AADN04002836.1       1555       1708  ENSGALG00000040165     .      +\n",
       "192  AADN04000891.1      22744      22863  ENSGALG00000029288     .      +\n",
       "193  AADN04005188.1      12747      12866  ENSGALG00000041326     .      +\n",
       "194  AADN04004449.1       9613       9732  ENSGALG00000043663     .      +\n",
       "195  AADN04009588.1       5404       5522  ENSGALG00000043431     .      +\n",
       "196  AADN04004449.1       1369       1488  ENSGALG00000040257     .      +\n",
       "197  AADN04003913.1      12476      12629  ENSGALG00000038196     .      -\n",
       "198  AADN04001404.1      23483      23602  ENSGALG00000031897     .      +\n",
       "199  AADN04003312.1       8596       8715  ENSGALG00000030604     .      +\n",
       "200  AADN04003312.1       2313       2366  ENSGALG00000031407     .      +\n",
       "201  AADN04011388.1       5207       5360  ENSGALG00000035877     .      -\n",
       "202  AADN04007387.1       7607       7756  ENSGALG00000036379     .      +\n",
       "\n",
       "[203 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.stop_codon.bed)>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for gene_id in get_gene_list(gene_dict):\n",
    "    start_codons = []\n",
    "    stop_codons = []\n",
    "    for start_codon in db.children(gene_id, featuretype='start_codon'):\n",
    "        ## 1 -based stop\n",
    "        ## 0-based start handled while converting to bed\n",
    "        start_codon.stop = start_codon.start\n",
    "        start_codons.append(start_codon)\n",
    "    for stop_codon in db.children(gene_id, featuretype='stop_codon'):\n",
    "        stop_codon.start = stop_codon.stop\n",
    "        stop_codon.stop = stop_codon.stop+1\n",
    "        stop_codons.append(stop_codon)\n",
    "    merged_start_codons = merge_regions(db, start_codons)\n",
    "    renamed_start_codons = rename_regions(merged_start_codons, gene_id)\n",
    "    merged_stop_codons = merge_regions(db, stop_codons)\n",
    "    renamed_stop_codons = rename_regions(merged_stop_codons, gene_id)\n",
    "    \n",
    "    start_codon_bed += create_bed(renamed_start_codons)    \n",
    "    stop_codon_bed += create_bed(renamed_stop_codons)\n",
    "\n",
    "\n",
    "    \n",
    "start_codon_bedtool = pybedtools.BedTool(start_codon_bed, from_string=True)\n",
    "stop_codon_bedtool = pybedtools.BedTool(stop_codon_bed, from_string=True)\n",
    "start_codon_bedtool.remove_invalid().sort().saveas('{}.start_codon.bed'.format(prefix))\n",
    "stop_codon_bedtool.remove_invalid().sort().saveas('{}.stop_codon.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:25:47.305834Z",
     "start_time": "2017-09-28T03:22:17.569193Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/annotation/Gallus_gallus.Gallus_gallus-5.0.91.gffutils.tss_sites.bed)>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TSS\n",
    "polyA_sites_bed = ''\n",
    "tss_sites_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    tss_sites = []\n",
    "    polyA_sites = []\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        start_t = copy.deepcopy(transcript)\n",
    "        stop_t = copy.deepcopy(transcript)\n",
    "        \n",
    "        start_t.stop = start_t.start + 1\n",
    "        \n",
    "        stop_t.start = stop_t.stop\n",
    "        \n",
    "        if transcript.strand == '-':\n",
    "            start_t, stop_t = stop_t, start_t\n",
    "        polyA_sites.append(start_t)\n",
    "        tss_sites.append(stop_t)\n",
    "    merged_polyA_sites = merge_regions(db, polyA_sites)\n",
    "    renamed_polyA_sites = rename_regions(merged_polyA_sites, gene_id)    \n",
    "    merged_tss_sites = merge_regions(db, tss_sites)\n",
    "    renamed_tss_sites = rename_regions(merged_tss_sites, gene_id)\n",
    "    polyA_sites_bed += create_bed(renamed_polyA_sites)    \n",
    "    tss_sites_bed += create_bed(renamed_tss_sites)\n",
    "\n",
    "polyA_sites_bedtool = pybedtools.BedTool(polyA_sites_bed, from_string=True)\n",
    "tss_sites_bedtool = pybedtools.BedTool(tss_sites_bed, from_string=True)\n",
    "polyA_sites_bedtool.remove_invalid().sort().saveas('{}.polyA_sites.bed'.format(prefix))\n",
    "tss_sites_bedtool.remove_invalid().sort().saveas('{}.tss_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:26:04.394168Z",
     "start_time": "2017-09-28T03:25:47.309124Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Genome file \"/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/fasta/Gallus_gallus.Gallus_gallus-5.0.91.dna.toplevel.sizes\" does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mValueError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-c8e3b035a63e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mtss\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtsses\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mas_bed6\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m#, merge_overlapping=True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mtss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove_invalid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msaveas\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{}.tss_temp.bed'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mpromoter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mslop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1000\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mg\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mchrsizes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mpromoter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove_invalid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msaveas\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'{}.promoter.1000.bed'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefix\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-panasas2/skchoudh/software_frozen/anaconda27/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mdecorated\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    804\u001b[0m             \u001b[0;31m# this calls the actual method in the first place; *result* is\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    805\u001b[0m             \u001b[0;31m# whatever you get back\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 806\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    808\u001b[0m             \u001b[0;31m# add appropriate tags\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-panasas2/skchoudh/software_frozen/anaconda27/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mwrapped\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    286\u001b[0m                             \u001b[0mcheck_for_genome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    287\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcheck_for_genome\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 288\u001b[0;31m                 \u001b[0mkwargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_genome\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m             \u001b[0;31m# For sequence methods, we may need to make a tempfile that will\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-panasas2/skchoudh/software_frozen/anaconda27/lib/python2.7/site-packages/pybedtools/bedtool.pyc\u001b[0m in \u001b[0;36mcheck_genome\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m   1448\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1449\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'g'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1450\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Genome file \"%s\" does not exist'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'g'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1451\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1452\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Genome file \"/home/cmb-panasas2/skchoudh/genomes/gallus_gallus_5.0/fasta/Gallus_gallus.Gallus_gallus-5.0.91.dna.toplevel.sizes\" does not exist"
     ]
    }
   ],
   "source": [
    "tss = tsses(db, as_bed6=True)#, merge_overlapping=True)\n",
    "tss.remove_invalid().sort().saveas('{}.tss_temp.bed'.format(prefix))\n",
    "promoter = tss.slop(l=1000, r=1000, s=True, g=chrsizes)\n",
    "promoter.remove_invalid().sort().saveas('{}.promoter.1000.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:26:12.000341Z",
     "start_time": "2017-09-28T03:26:04.398959Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "for l in [1000, 2000, 3000, 4000, 5000]:\n",
    "    promoter = tss.slop(l=l, r=l, s=True, g=chrsizes)\n",
    "    promoter.remove_invalid().sort().saveas('{}.promoter.{}.bed'.format(prefix, l))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "12px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
