{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:08:09.166605Z",
     "start_time": "2017-09-28T03:08:07.636901Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict, OrderedDict\n",
    "import warnings\n",
    "import logging\n",
    "import gffutils\n",
    "import pybedtools\n",
    "import pandas as pd\n",
    "import copy\n",
    "import re\n",
    "from gffutils.pybedtools_integration import tsses\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:11:54.277345Z",
     "start_time": "2017-09-28T03:11:54.265908Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "gtf = '/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gtf'\n",
    "gtf_db = '/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gtf.db'\n",
    "prefix = '/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils'\n",
    "chrsizes = '/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/fasta/Xenopus_tropicalis.JGI_4.2.dna.toplevel.sizes'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:15:43.298837Z",
     "start_time": "2017-09-28T03:11:55.256580Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.create_db(gtf, dbfn=gtf_db, disable_infer_genes=True, disable_infer_transcripts=True, merge_strategy='merge', force=True)\n",
    "def create_gene_dict(db):\n",
    "    '''\n",
    "    Store each feature line db.all_features() as a dict of dicts\n",
    "    '''\n",
    "    gene_dict = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))\n",
    "    for line_no, feature in enumerate(db.all_features()):\n",
    "        gene_ids = feature.attributes['gene_id']\n",
    "        feature_type = feature.featuretype\n",
    "        if feature_type == 'gene':\n",
    "            if len(gene_ids)!=1:\n",
    "                logging.warning('Found multiple gene_ids on line {} in gtf'.format(line_no))\n",
    "                break\n",
    "            else:\n",
    "                gene_id = gene_ids[0]\n",
    "                gene_dict[gene_id]['gene'] = feature\n",
    "        else:\n",
    "            transcript_ids = feature.attributes['transcript_id']\n",
    "\n",
    "            for gene_id in gene_ids:\n",
    "                for transcript_id in transcript_ids:\n",
    "                    gene_dict[gene_id][transcript_id][feature_type].append(feature)\n",
    "    return gene_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.280784Z",
     "start_time": "2017-09-28T03:15:43.304709Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(gtf_db, keep_order=True)\n",
    "gene_dict = create_gene_dict(db)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.353531Z",
     "start_time": "2017-09-28T03:16:43.285946Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS\n",
      "exon\n",
      "five_prime_utr\n",
      "gene\n",
      "start_codon\n",
      "stop_codon\n",
      "three_prime_utr\n",
      "transcript\n"
     ]
    }
   ],
   "source": [
    "for x in db.featuretypes():\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:16:43.461123Z",
     "start_time": "2017-09-28T03:16:43.357554Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "def get_gene_list(gene_dict):\n",
    "    return list(set(gene_dict.keys()))\n",
    "\n",
    "def get_UTR_regions(gene_dict, gene_id, transcript, cds):\n",
    "    if len(cds)==0:\n",
    "        return [], []\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    utrs = gene_dict[gene_id][transcript]['UTR']\n",
    "    first_cds = cds[0]\n",
    "    last_cds = cds[-1]\n",
    "    for utr in utrs:\n",
    "        ## Push all cds at once\n",
    "        ## Sort later to remove duplicates\n",
    "        strand = utr.strand\n",
    "        if strand == '+':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')\n",
    "        elif strand == '-':\n",
    "            if utr.stop < first_cds.start:\n",
    "                utr.feature_type = 'three_prime_UTR'\n",
    "                utr3_regions.append(utr)\n",
    "            elif utr.start > last_cds.stop:\n",
    "                utr.feature_type = 'five_prime_UTR'\n",
    "                utr5_regions.append(utr)                \n",
    "            else:\n",
    "                raise RuntimeError('Error with cds')    \n",
    "    return utr5_regions, utr3_regions\n",
    "    \n",
    "def create_bed(regions, bedtype='0'):\n",
    "    '''Create bed from list of regions\n",
    "    bedtype: 0 or 1\n",
    "        0-Based or 1-based coordinate of the BED\n",
    "    '''\n",
    "    bedstr = ''\n",
    "    for region in regions:\n",
    "        assert len(region.attributes['gene_id']) == 1\n",
    "        ## GTF start is 1-based, so shift by one while writing \n",
    "        ## to 0-based BED format\n",
    "        if bedtype == '0':\n",
    "            start = region.start - 1\n",
    "        else:\n",
    "            start = region.start\n",
    "        bedstr += '{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n'.format(region.chrom,\n",
    "                                             start,\n",
    "                                             region.stop,\n",
    "                                             re.sub('\\.\\d+', '', region.attributes['gene_id'][0]),\n",
    "                                             '.',\n",
    "                                             region.strand)\n",
    "    # Remove duplicates\n",
    "    bedstr = '\\n'.join(list(OrderedDict.fromkeys(bedstr.split('\\n'))))\n",
    "    return bedstr\n",
    "\n",
    "def rename_regions(regions, gene_id):\n",
    "    regions = list(regions)\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    for region in regions:\n",
    "        region.attributes['gene_id'] = gene_id\n",
    "    return regions\n",
    "\n",
    "def merge_regions(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start))\n",
    "    return merged\n",
    "\n",
    "def merge_regions_nostrand(db, regions):\n",
    "    if len(regions) == 0:\n",
    "        return []\n",
    "    merged = db.merge(sorted(list(regions), key=lambda x: x.start), ignore_strand=True)\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:17:38.562657Z",
     "start_time": "2017-09-28T03:16:43.463672Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.cds.bed)>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_bed = ''\n",
    "utr3_bed = ''\n",
    "\n",
    "\n",
    "\n",
    "gene_bed = ''\n",
    "exon_bed = ''\n",
    "intron_bed = ''\n",
    "start_codon_bed = ''\n",
    "stop_codon_bed = ''\n",
    "cds_bed = ''\n",
    "\n",
    "gene_list = []\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    gene_list.append(gene_dict[gene_id]['gene'])\n",
    "    \n",
    "    utr5_regions, utr3_regions = [], []\n",
    "    exon_regions, intron_regions = [], []\n",
    "    star_codon_regions, stop_codon_regions = [], []\n",
    "    cds_regions = []\n",
    "    \n",
    "    for feature in gene_dict[gene_id].keys():\n",
    "        if feature == 'gene':\n",
    "            continue\n",
    "        cds = list(gene_dict[gene_id][feature]['CDS'])\n",
    "        exons = list(gene_dict[gene_id][feature]['exon'])\n",
    "        merged_exons = merge_regions(db, exons)\n",
    "        introns = db.interfeatures(merged_exons)\n",
    "        utr5_region = list(gene_dict[gene_id][feature]['five_prime_utr'])\n",
    "        utr3_region = list(gene_dict[gene_id][feature]['three_prime_utr'])\n",
    "        utr5_regions += utr5_region\n",
    "        utr3_regions += utr3_region\n",
    "        exon_regions += exons\n",
    "        intron_regions += introns\n",
    "        cds_regions += cds\n",
    "        \n",
    "    merged_utr5 = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5 = rename_regions(merged_utr5, gene_id)\n",
    "    \n",
    "    merged_utr3 = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3 = rename_regions(merged_utr3, gene_id)\n",
    "    \n",
    "    merged_exons = merge_regions(db, exon_regions)\n",
    "    renamed_exons = rename_regions(merged_exons, gene_id)\n",
    "    \n",
    "    merged_introns = merge_regions(db, intron_regions)\n",
    "    renamed_introns = rename_regions(merged_introns, gene_id)\n",
    "    \n",
    "    merged_cds = merge_regions(db, cds_regions)\n",
    "    renamed_cds = rename_regions(merged_cds, gene_id)\n",
    "    \n",
    "    utr3_bed += create_bed(renamed_utr3)\n",
    "    utr5_bed += create_bed(renamed_utr5)\n",
    "    exon_bed += create_bed(renamed_exons)\n",
    "    intron_bed += create_bed(renamed_introns)\n",
    "    cds_bed += create_bed(renamed_cds)\n",
    "    \n",
    "    \n",
    "gene_bed = create_bed(gene_list)\n",
    "gene_bedtool = pybedtools.BedTool(gene_bed, from_string=True)\n",
    "utr5_bedtool = pybedtools.BedTool(utr5_bed, from_string=True)\n",
    "utr3_bedtool = pybedtools.BedTool(utr3_bed, from_string=True)\n",
    "exon_bedtool = pybedtools.BedTool(exon_bed, from_string=True)\n",
    "intron_bedtool = pybedtools.BedTool(intron_bed, from_string=True)\n",
    "cds_bedtool = pybedtools.BedTool(cds_bed, from_string=True)\n",
    "\n",
    "gene_bedtool.remove_invalid().sort().saveas('{}.genes.bed'.format(prefix))\n",
    "utr5_bedtool.remove_invalid().sort().saveas('{}.UTR5.bed'.format(prefix))\n",
    "utr3_bedtool.remove_invalid().sort().saveas('{}.UTR3.bed'.format(prefix))\n",
    "exon_bedtool.remove_invalid().sort().saveas('{}.exon.bed'.format(prefix))\n",
    "intron_bedtool.remove_invalid().sort().saveas('{}.intron.bed'.format(prefix))\n",
    "cds_bedtool.remove_invalid().sort().saveas('{}.cds.bed'.format(prefix))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T04:25:18.715414Z",
     "start_time": "2017-09-28T04:20:10.158492Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "utr5_region_bed = ''\n",
    "utr3_region_bed = ''\n",
    "\n",
    "\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    utr5_regions = []\n",
    "    utr3_regions = []\n",
    "    for utr5_region in db.children(gene_id, featuretype='five_prime_utr'):\n",
    "        utr5_regions.append(utr5_region)\n",
    "    for utr3_region in db.children(gene_id, featuretype='three_prime_utr'):\n",
    "        utr3_regions.append(utr3_region)\n",
    "    merged_utr5_regions = merge_regions(db, utr5_regions)\n",
    "    renamed_utr5_regions = rename_regions(merged_utr5_regions, gene_id)\n",
    "    merged_utr3_regions = merge_regions(db, utr3_regions)\n",
    "    renamed_utr3_regions = rename_regions(merged_utr3_regions, gene_id)\n",
    "    \n",
    "    utr5_region_bed += create_bed(renamed_utr5_regions)    \n",
    "    utr3_region_bed += create_bed(renamed_utr3_regions)\n",
    "\n",
    "\n",
    "    \n",
    "utr5_region_bedtool = pybedtools.BedTool(utr5_region_bed, from_string=True)\n",
    "utr3_region_bedtool = pybedtools.BedTool(utr3_region_bed, from_string=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:08:46.134982Z",
     "start_time": "2017-09-28T05:08:45.236263Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr3_region.bed)>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr5_region_bedtool.remove_invalid().sort().saveas('{}.utr5_region.bed'.format(prefix))\n",
    "utr3_region_bedtool.remove_invalid().sort().saveas('{}.utr3_region.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:09:02.138861Z",
     "start_time": "2017-09-28T05:09:02.130605Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr5_region.bed'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'{}.utr5_region.bed'.format(prefix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T05:09:21.627640Z",
     "start_time": "2017-09-28T05:09:21.075155Z"
    }
   },
   "outputs": [],
   "source": [
    "!bedSort /home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr3_region.bed /home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr3_region.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!bedSort /home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr5_region.bed /home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.utr5_region.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.tRNA_sites.bed)>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites = []\n",
    "tRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'tRNA' in transcript.attributes['gene_biotype'] or 'Mt_tRNA' in transcript.attributes['transcript_biotype']:\n",
    "            tRNA_sites.append(transcript)\n",
    "    #merged_tRNA_sites = merge_regions_nostrand(db, tRNA_sites)\n",
    "    #renamed_tRNA_sites = rename_regions(merged_tRNA_sites, gene_id)\n",
    "    tRNA_bed += create_bed(tRNA_sites)\n",
    "\n",
    "tRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(tRNA_bed.split('\\n'))))\n",
    "tRNA_sites_bedtool = pybedtools.BedTool(tRNA_bed, from_string=True)\n",
    "tRNA_sites_bedtool.remove_invalid().sort().saveas('{}.tRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>9498</td>\n",
       "      <td>9568</td>\n",
       "      <td>ENSXETG00000034379</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>5004</td>\n",
       "      <td>5073</td>\n",
       "      <td>ENSXETG00000034366</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>11648</td>\n",
       "      <td>11717</td>\n",
       "      <td>ENSXETG00000034384</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>11717</td>\n",
       "      <td>11785</td>\n",
       "      <td>ENSXETG00000034385</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>11785</td>\n",
       "      <td>11858</td>\n",
       "      <td>ENSXETG00000034386</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>9911</td>\n",
       "      <td>9980</td>\n",
       "      <td>ENSXETG00000034381</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>14186</td>\n",
       "      <td>14255</td>\n",
       "      <td>ENSXETG00000034389</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>7798</td>\n",
       "      <td>7872</td>\n",
       "      <td>ENSXETG00000034375</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>7039</td>\n",
       "      <td>7108</td>\n",
       "      <td>ENSXETG00000034373</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>6953</td>\n",
       "      <td>7024</td>\n",
       "      <td>ENSXETG00000034372</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>3759</td>\n",
       "      <td>3830</td>\n",
       "      <td>ENSXETG00000034362</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>3829</td>\n",
       "      <td>3900</td>\n",
       "      <td>ENSXETG00000034363</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>2716</td>\n",
       "      <td>2791</td>\n",
       "      <td>ENSXETG00000034360</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>5146</td>\n",
       "      <td>5219</td>\n",
       "      <td>ENSXETG00000034368</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>5257</td>\n",
       "      <td>5323</td>\n",
       "      <td>ENSXETG00000034369</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>15499</td>\n",
       "      <td>15566</td>\n",
       "      <td>ENSXETG00000034392</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>15399</td>\n",
       "      <td>15470</td>\n",
       "      <td>ENSXETG00000034391</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>ENSXETG00000034356</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>5076</td>\n",
       "      <td>5145</td>\n",
       "      <td>ENSXETG00000034367</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>3899</td>\n",
       "      <td>3968</td>\n",
       "      <td>ENSXETG00000034364</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>5323</td>\n",
       "      <td>5393</td>\n",
       "      <td>ENSXETG00000034370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NC_006839</td>\n",
       "      <td>1011</td>\n",
       "      <td>1081</td>\n",
       "      <td>ENSXETG00000034358</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom  start    end                name score strand\n",
       "0   NC_006839   9498   9568  ENSXETG00000034379     .      +\n",
       "1   NC_006839   5004   5073  ENSXETG00000034366     .      +\n",
       "2   NC_006839  11648  11717  ENSXETG00000034384     .      +\n",
       "3   NC_006839  11717  11785  ENSXETG00000034385     .      +\n",
       "4   NC_006839  11785  11858  ENSXETG00000034386     .      +\n",
       "5   NC_006839   9911   9980  ENSXETG00000034381     .      +\n",
       "6   NC_006839  14186  14255  ENSXETG00000034389     .      -\n",
       "7   NC_006839   7798   7872  ENSXETG00000034375     .      +\n",
       "8   NC_006839   7039   7108  ENSXETG00000034373     .      +\n",
       "9   NC_006839   6953   7024  ENSXETG00000034372     .      -\n",
       "10  NC_006839   3759   3830  ENSXETG00000034362     .      +\n",
       "11  NC_006839   3829   3900  ENSXETG00000034363     .      -\n",
       "12  NC_006839   2716   2791  ENSXETG00000034360     .      +\n",
       "13  NC_006839   5146   5219  ENSXETG00000034368     .      -\n",
       "14  NC_006839   5257   5323  ENSXETG00000034369     .      -\n",
       "15  NC_006839  15499  15566  ENSXETG00000034392     .      +\n",
       "16  NC_006839  15399  15470  ENSXETG00000034391     .      +\n",
       "17  NC_006839      0     68  ENSXETG00000034356     .      +\n",
       "18  NC_006839   5076   5145  ENSXETG00000034367     .      -\n",
       "19  NC_006839   3899   3968  ENSXETG00000034364     .      +\n",
       "20  NC_006839   5323   5393  ENSXETG00000034370     .      -\n",
       "21  NC_006839   1011   1081  ENSXETG00000034358     .      +"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.rRNA_sites.bed)>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites = []\n",
    "rRNA_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        if 'rRNA' in transcript.attributes['gene_biotype']:\n",
    "            rRNA_sites.append(transcript)\n",
    "    #renamed_rRNA_sites = rename_regions(rRNA_sites, gene_id)\n",
    "    rRNA_bed += create_bed(rRNA_sites)\n",
    "rRNA_bed = '\\n'.join(list(OrderedDict.fromkeys(rRNA_bed.split('\\n'))))\n",
    "rRNA_sites_bedtool = pybedtools.BedTool(rRNA_bed, from_string=True)\n",
    "rRNA_sites_bedtool.remove_invalid().sort().saveas('{}.rRNA_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL173208.1</td>\n",
       "      <td>205449</td>\n",
       "      <td>205579</td>\n",
       "      <td>ENSXETG00000029707</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL172919.1</td>\n",
       "      <td>1153290</td>\n",
       "      <td>1153409</td>\n",
       "      <td>ENSXETG00000029943</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL189750.1</td>\n",
       "      <td>973</td>\n",
       "      <td>1092</td>\n",
       "      <td>ENSXETG00000031678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL173818.1</td>\n",
       "      <td>75436</td>\n",
       "      <td>75555</td>\n",
       "      <td>ENSXETG00000034259</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL177336.1</td>\n",
       "      <td>7772</td>\n",
       "      <td>7885</td>\n",
       "      <td>ENSXETG00000031536</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GL172919.1</td>\n",
       "      <td>1153575</td>\n",
       "      <td>1153610</td>\n",
       "      <td>ENSXETG00000029722</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GL173415.1</td>\n",
       "      <td>198515</td>\n",
       "      <td>198633</td>\n",
       "      <td>ENSXETG00000029724</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GL173816.1</td>\n",
       "      <td>2970</td>\n",
       "      <td>3070</td>\n",
       "      <td>ENSXETG00000028457</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>57950</td>\n",
       "      <td>58069</td>\n",
       "      <td>ENSXETG00000028456</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>40568</td>\n",
       "      <td>40687</td>\n",
       "      <td>ENSXETG00000033568</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GL173381.1</td>\n",
       "      <td>110724</td>\n",
       "      <td>110838</td>\n",
       "      <td>ENSXETG00000031885</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GL173707.1</td>\n",
       "      <td>183537</td>\n",
       "      <td>183656</td>\n",
       "      <td>ENSXETG00000031550</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GL175741.1</td>\n",
       "      <td>8540</td>\n",
       "      <td>8659</td>\n",
       "      <td>ENSXETG00000032288</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GL173818.1</td>\n",
       "      <td>68334</td>\n",
       "      <td>68453</td>\n",
       "      <td>ENSXETG00000033770</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GL189750.1</td>\n",
       "      <td>445</td>\n",
       "      <td>564</td>\n",
       "      <td>ENSXETG00000033952</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GL189750.1</td>\n",
       "      <td>709</td>\n",
       "      <td>828</td>\n",
       "      <td>ENSXETG00000030279</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GL174202.1</td>\n",
       "      <td>69545</td>\n",
       "      <td>69664</td>\n",
       "      <td>ENSXETG00000028478</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GL172639.1</td>\n",
       "      <td>3585760</td>\n",
       "      <td>3585878</td>\n",
       "      <td>ENSXETG00000029301</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GL173459.1</td>\n",
       "      <td>204775</td>\n",
       "      <td>204894</td>\n",
       "      <td>ENSXETG00000029304</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GL173797.1</td>\n",
       "      <td>59573</td>\n",
       "      <td>59692</td>\n",
       "      <td>ENSXETG00000032947</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GL172660.1</td>\n",
       "      <td>771879</td>\n",
       "      <td>771976</td>\n",
       "      <td>ENSXETG00000029685</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GL172660.1</td>\n",
       "      <td>844146</td>\n",
       "      <td>844224</td>\n",
       "      <td>ENSXETG00000029689</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>46191</td>\n",
       "      <td>46310</td>\n",
       "      <td>ENSXETG00000031717</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GL172919.1</td>\n",
       "      <td>1157248</td>\n",
       "      <td>1157367</td>\n",
       "      <td>ENSXETG00000030619</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GL174641.1</td>\n",
       "      <td>712</td>\n",
       "      <td>831</td>\n",
       "      <td>ENSXETG00000031409</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>40832</td>\n",
       "      <td>40951</td>\n",
       "      <td>ENSXETG00000031388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GL173707.1</td>\n",
       "      <td>186719</td>\n",
       "      <td>186838</td>\n",
       "      <td>ENSXETG00000031400</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GL190802.1</td>\n",
       "      <td>87</td>\n",
       "      <td>206</td>\n",
       "      <td>ENSXETG00000033034</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GL172660.1</td>\n",
       "      <td>30535</td>\n",
       "      <td>30660</td>\n",
       "      <td>ENSXETG00000029663</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GL178569.1</td>\n",
       "      <td>2918</td>\n",
       "      <td>3037</td>\n",
       "      <td>ENSXETG00000028202</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>GL173797.1</td>\n",
       "      <td>55015</td>\n",
       "      <td>55134</td>\n",
       "      <td>ENSXETG00000030658</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>GL182039.1</td>\n",
       "      <td>3291</td>\n",
       "      <td>3410</td>\n",
       "      <td>ENSXETG00000032777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>GL172831.1</td>\n",
       "      <td>790430</td>\n",
       "      <td>790546</td>\n",
       "      <td>ENSXETG00000029747</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>GL174641.1</td>\n",
       "      <td>1235</td>\n",
       "      <td>1354</td>\n",
       "      <td>ENSXETG00000030186</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>GL182039.1</td>\n",
       "      <td>2646</td>\n",
       "      <td>2765</td>\n",
       "      <td>ENSXETG00000030187</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>GL189750.1</td>\n",
       "      <td>191</td>\n",
       "      <td>300</td>\n",
       "      <td>ENSXETG00000031329</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>GL173409.1</td>\n",
       "      <td>293325</td>\n",
       "      <td>293444</td>\n",
       "      <td>ENSXETG00000031324</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>GL173797.1</td>\n",
       "      <td>60119</td>\n",
       "      <td>60238</td>\n",
       "      <td>ENSXETG00000033670</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>GL173816.1</td>\n",
       "      <td>8726</td>\n",
       "      <td>8835</td>\n",
       "      <td>ENSXETG00000028767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>GL190659.1</td>\n",
       "      <td>106</td>\n",
       "      <td>225</td>\n",
       "      <td>ENSXETG00000033890</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>GL172919.1</td>\n",
       "      <td>1156467</td>\n",
       "      <td>1156576</td>\n",
       "      <td>ENSXETG00000028189</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>42152</td>\n",
       "      <td>42271</td>\n",
       "      <td>ENSXETG00000033330</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>GL173797.1</td>\n",
       "      <td>55279</td>\n",
       "      <td>55398</td>\n",
       "      <td>ENSXETG00000031583</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>GL173818.1</td>\n",
       "      <td>74397</td>\n",
       "      <td>74516</td>\n",
       "      <td>ENSXETG00000033798</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>GL173816.1</td>\n",
       "      <td>7446</td>\n",
       "      <td>7565</td>\n",
       "      <td>ENSXETG00000033422</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>GL173707.1</td>\n",
       "      <td>188644</td>\n",
       "      <td>188763</td>\n",
       "      <td>ENSXETG00000030219</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>GL190659.1</td>\n",
       "      <td>370</td>\n",
       "      <td>489</td>\n",
       "      <td>ENSXETG00000030215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>GL190802.1</td>\n",
       "      <td>351</td>\n",
       "      <td>470</td>\n",
       "      <td>ENSXETG00000033937</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>GL174142.1</td>\n",
       "      <td>43815</td>\n",
       "      <td>43934</td>\n",
       "      <td>ENSXETG00000033935</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>GL176103.1</td>\n",
       "      <td>7975</td>\n",
       "      <td>8094</td>\n",
       "      <td>ENSXETG00000028141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>GL174641.1</td>\n",
       "      <td>976</td>\n",
       "      <td>1095</td>\n",
       "      <td>ENSXETG00000032309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>GL176103.1</td>\n",
       "      <td>7721</td>\n",
       "      <td>7830</td>\n",
       "      <td>ENSXETG00000028777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262</th>\n",
       "      <td>GL183261.1</td>\n",
       "      <td>903</td>\n",
       "      <td>1022</td>\n",
       "      <td>ENSXETG00000033654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>GL174202.1</td>\n",
       "      <td>70602</td>\n",
       "      <td>70721</td>\n",
       "      <td>ENSXETG00000030086</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>GL172974.1</td>\n",
       "      <td>672068</td>\n",
       "      <td>672185</td>\n",
       "      <td>ENSXETG00000028709</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>GL173707.1</td>\n",
       "      <td>187948</td>\n",
       "      <td>188067</td>\n",
       "      <td>ENSXETG00000032468</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>GL178992.1</td>\n",
       "      <td>2487</td>\n",
       "      <td>2641</td>\n",
       "      <td>ENSXETG00000030685</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>GL188195.1</td>\n",
       "      <td>18</td>\n",
       "      <td>137</td>\n",
       "      <td>ENSXETG00000030020</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>GL173647.1</td>\n",
       "      <td>100514</td>\n",
       "      <td>100633</td>\n",
       "      <td>ENSXETG00000028112</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>GL178273.1</td>\n",
       "      <td>3967</td>\n",
       "      <td>4119</td>\n",
       "      <td>ENSXETG00000032797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>270 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom    start      end                name score strand\n",
       "0    GL173208.1   205449   205579  ENSXETG00000029707     .      -\n",
       "1    GL172919.1  1153290  1153409  ENSXETG00000029943     .      +\n",
       "2    GL189750.1      973     1092  ENSXETG00000031678     .      -\n",
       "3    GL173818.1    75436    75555  ENSXETG00000034259     .      -\n",
       "4    GL177336.1     7772     7885  ENSXETG00000031536     .      -\n",
       "5    GL172919.1  1153575  1153610  ENSXETG00000029722     .      +\n",
       "6    GL173415.1   198515   198633  ENSXETG00000029724     .      -\n",
       "7    GL173816.1     2970     3070  ENSXETG00000028457     .      +\n",
       "8    GL174142.1    57950    58069  ENSXETG00000028456     .      +\n",
       "9    GL174142.1    40568    40687  ENSXETG00000033568     .      +\n",
       "10   GL173381.1   110724   110838  ENSXETG00000031885     .      -\n",
       "11   GL173707.1   183537   183656  ENSXETG00000031550     .      -\n",
       "12   GL175741.1     8540     8659  ENSXETG00000032288     .      -\n",
       "13   GL173818.1    68334    68453  ENSXETG00000033770     .      -\n",
       "14   GL189750.1      445      564  ENSXETG00000033952     .      -\n",
       "15   GL189750.1      709      828  ENSXETG00000030279     .      -\n",
       "16   GL174202.1    69545    69664  ENSXETG00000028478     .      -\n",
       "17   GL172639.1  3585760  3585878  ENSXETG00000029301     .      +\n",
       "18   GL173459.1   204775   204894  ENSXETG00000029304     .      -\n",
       "19   GL173797.1    59573    59692  ENSXETG00000032947     .      -\n",
       "20   GL172660.1   771879   771976  ENSXETG00000029685     .      +\n",
       "21   GL172660.1   844146   844224  ENSXETG00000029689     .      +\n",
       "22   GL174142.1    46191    46310  ENSXETG00000031717     .      +\n",
       "23   GL172919.1  1157248  1157367  ENSXETG00000030619     .      +\n",
       "24   GL174641.1      712      831  ENSXETG00000031409     .      -\n",
       "25   GL174142.1    40832    40951  ENSXETG00000031388     .      +\n",
       "26   GL173707.1   186719   186838  ENSXETG00000031400     .      -\n",
       "27   GL190802.1       87      206  ENSXETG00000033034     .      +\n",
       "28   GL172660.1    30535    30660  ENSXETG00000029663     .      -\n",
       "29   GL178569.1     2918     3037  ENSXETG00000028202     .      -\n",
       "..          ...      ...      ...                 ...   ...    ...\n",
       "240  GL173797.1    55015    55134  ENSXETG00000030658     .      -\n",
       "241  GL182039.1     3291     3410  ENSXETG00000032777     .      -\n",
       "242  GL172831.1   790430   790546  ENSXETG00000029747     .      -\n",
       "243  GL174641.1     1235     1354  ENSXETG00000030186     .      -\n",
       "244  GL182039.1     2646     2765  ENSXETG00000030187     .      -\n",
       "245  GL189750.1      191      300  ENSXETG00000031329     .      -\n",
       "246  GL173409.1   293325   293444  ENSXETG00000031324     .      +\n",
       "247  GL173797.1    60119    60238  ENSXETG00000033670     .      -\n",
       "248  GL173816.1     8726     8835  ENSXETG00000028767     .      +\n",
       "249  GL190659.1      106      225  ENSXETG00000033890     .      -\n",
       "250  GL172919.1  1156467  1156576  ENSXETG00000028189     .      +\n",
       "251  GL174142.1    42152    42271  ENSXETG00000033330     .      +\n",
       "252  GL173797.1    55279    55398  ENSXETG00000031583     .      -\n",
       "253  GL173818.1    74397    74516  ENSXETG00000033798     .      -\n",
       "254  GL173816.1     7446     7565  ENSXETG00000033422     .      +\n",
       "255  GL173707.1   188644   188763  ENSXETG00000030219     .      -\n",
       "256  GL190659.1      370      489  ENSXETG00000030215     .      -\n",
       "257  GL190802.1      351      470  ENSXETG00000033937     .      +\n",
       "258  GL174142.1    43815    43934  ENSXETG00000033935     .      +\n",
       "259  GL176103.1     7975     8094  ENSXETG00000028141     .      -\n",
       "260  GL174641.1      976     1095  ENSXETG00000032309     .      -\n",
       "261  GL176103.1     7721     7830  ENSXETG00000028777     .      -\n",
       "262  GL183261.1      903     1022  ENSXETG00000033654     .      -\n",
       "263  GL174202.1    70602    70721  ENSXETG00000030086     .      -\n",
       "264  GL172974.1   672068   672185  ENSXETG00000028709     .      +\n",
       "265  GL173707.1   187948   188067  ENSXETG00000032468     .      -\n",
       "266  GL178992.1     2487     2641  ENSXETG00000030685     .      -\n",
       "267  GL188195.1       18      137  ENSXETG00000030020     .      -\n",
       "268  GL173647.1   100514   100633  ENSXETG00000028112     .      -\n",
       "269  GL178273.1     3967     4119  ENSXETG00000032797     .      -\n",
       "\n",
       "[270 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_sites_bedtool.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.stop_codon.bed)>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for gene_id in get_gene_list(gene_dict):\n",
    "    start_codons = []\n",
    "    stop_codons = []\n",
    "    for start_codon in db.children(gene_id, featuretype='start_codon'):\n",
    "        ## 1 -based stop\n",
    "        ## 0-based start handled while converting to bed\n",
    "        start_codon.stop = start_codon.start\n",
    "        start_codons.append(start_codon)\n",
    "    for stop_codon in db.children(gene_id, featuretype='stop_codon'):\n",
    "        stop_codon.start = stop_codon.stop\n",
    "        stop_codon.stop = stop_codon.stop+1\n",
    "        stop_codons.append(stop_codon)\n",
    "    merged_start_codons = merge_regions(db, start_codons)\n",
    "    renamed_start_codons = rename_regions(merged_start_codons, gene_id)\n",
    "    merged_stop_codons = merge_regions(db, stop_codons)\n",
    "    renamed_stop_codons = rename_regions(merged_stop_codons, gene_id)\n",
    "    \n",
    "    start_codon_bed += create_bed(renamed_start_codons)    \n",
    "    stop_codon_bed += create_bed(renamed_stop_codons)\n",
    "\n",
    "\n",
    "    \n",
    "start_codon_bedtool = pybedtools.BedTool(start_codon_bed, from_string=True)\n",
    "stop_codon_bedtool = pybedtools.BedTool(stop_codon_bed, from_string=True)\n",
    "start_codon_bedtool.remove_invalid().sort().saveas('{}.start_codon.bed'.format(prefix))\n",
    "stop_codon_bedtool.remove_invalid().sort().saveas('{}.stop_codon.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:25:47.305834Z",
     "start_time": "2017-09-28T03:22:17.569193Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.tss_sites.bed)>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TSS\n",
    "polyA_sites_bed = ''\n",
    "tss_sites_bed = ''\n",
    "for gene_id in get_gene_list(gene_dict):\n",
    "    tss_sites = []\n",
    "    polyA_sites = []\n",
    "    for transcript in db.children(gene_id, featuretype='transcript'):\n",
    "        start_t = copy.deepcopy(transcript)\n",
    "        stop_t = copy.deepcopy(transcript)\n",
    "        \n",
    "        start_t.stop = start_t.start + 1\n",
    "        \n",
    "        stop_t.start = stop_t.stop\n",
    "        \n",
    "        if transcript.strand == '-':\n",
    "            start_t, stop_t = stop_t, start_t\n",
    "        polyA_sites.append(start_t)\n",
    "        tss_sites.append(stop_t)\n",
    "    merged_polyA_sites = merge_regions(db, polyA_sites)\n",
    "    renamed_polyA_sites = rename_regions(merged_polyA_sites, gene_id)    \n",
    "    merged_tss_sites = merge_regions(db, tss_sites)\n",
    "    renamed_tss_sites = rename_regions(merged_tss_sites, gene_id)\n",
    "    polyA_sites_bed += create_bed(renamed_polyA_sites)    \n",
    "    tss_sites_bed += create_bed(renamed_tss_sites)\n",
    "\n",
    "polyA_sites_bedtool = pybedtools.BedTool(polyA_sites_bed, from_string=True)\n",
    "tss_sites_bedtool = pybedtools.BedTool(tss_sites_bed, from_string=True)\n",
    "polyA_sites_bedtool.remove_invalid().sort().saveas('{}.polyA_sites.bed'.format(prefix))\n",
    "tss_sites_bedtool.remove_invalid().sort().saveas('{}.tss_sites.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:26:04.394168Z",
     "start_time": "2017-09-28T03:25:47.309124Z"
    },
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(/home/cmb-panasas2/skchoudh/genomes/xenopus_tropicalis_JGI_4.2/annotation/Xenopus_tropicalis.JGI_4.2.91.gffutils.promoter.1000.bed)>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss = tsses(db, as_bed6=True)#, merge_overlapping=True)\n",
    "tss.remove_invalid().sort().saveas('{}.tss_temp.bed'.format(prefix))\n",
    "promoter = tss.slop(l=1000, r=1000, s=True, g=chrsizes)\n",
    "promoter.remove_invalid().sort().saveas('{}.promoter.1000.bed'.format(prefix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-28T03:26:12.000341Z",
     "start_time": "2017-09-28T03:26:04.398959Z"
    },
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "for l in [1000, 2000, 3000, 4000, 5000]:\n",
    "    promoter = tss.slop(l=l, r=l, s=True, g=chrsizes)\n",
    "    promoter.remove_invalid().sort().saveas('{}.promoter.{}.bed'.format(prefix, l))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "12px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": true,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
