{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import pandas as pd\n",
    "import pyfaidx\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "five_prime_fasta = '/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/SGD_all_ORFs/SGD_all_ORFs_5prime_UTRs.fsa'\n",
    "three_prime_fasta = '/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/SGD_all_ORFs/SGD_all_ORFs_3prime_UTRs.fsa'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "FastaIndexingError",
     "evalue": "Line length of fasta file is not consistent! Inconsistent line found in >sacCer3_ct_UserTrack_3545_YAL055W_SC_Transcript_00009455_Miura_2006 at line 21.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFastaIndexingError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-fed57beb4bd2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfive_prime_idx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpyfaidx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFasta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfive_prime_fasta\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mthree_prime_idx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpyfaidx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFasta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mthree_prime_fasta\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-06/as/skchoudh/software_frozen/anaconda37/envs/riboraptor/lib/python3.7/site-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, filt_function, one_based_attributes, read_long_names, duplicate_action, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    994\u001b[0m             \u001b[0msequence_always_upper\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msequence_always_upper\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    995\u001b[0m             \u001b[0mrebuild\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrebuild\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 996\u001b[0;31m             build_index=build_index)\n\u001b[0m\u001b[1;32m    997\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfaidx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    998\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmutable\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-06/as/skchoudh/software_frozen/anaconda37/envs/riboraptor/lib/python3.7/site-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, default_seq, key_function, as_raw, strict_bounds, read_ahead, mutable, split_char, duplicate_action, filt_function, one_based_attributes, read_long_names, sequence_always_upper, rebuild, build_index)\u001b[0m\n\u001b[1;32m    421\u001b[0m                             self.indexname, self.filename), RuntimeWarning)\n\u001b[1;32m    422\u001b[0m                 \u001b[0;32melif\u001b[0m \u001b[0mbuild_index\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 423\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuild_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    424\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_fai\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    425\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-06/as/skchoudh/software_frozen/anaconda37/envs/riboraptor/lib/python3.7/site-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36mbuild_index\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    592\u001b[0m                     % self.indexname)\n\u001b[1;32m    593\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mFastaIndexingError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 594\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    595\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    596\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite_fai\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/cmb-06/as/skchoudh/software_frozen/anaconda37/envs/riboraptor/lib/python3.7/site-packages/pyfaidx/__init__.py\u001b[0m in \u001b[0;36mbuild_index\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    535\u001b[0m                                     \u001b[0;34m\"Inconsistent line found in >{0} at \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    536\u001b[0m                                     \"line {1:n}.\".format(\n\u001b[0;32m--> 537\u001b[0;31m                                         rname, bad_lines[0][0] + 1))\n\u001b[0m\u001b[1;32m    538\u001b[0m                             \u001b[0mblen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    539\u001b[0m                             \u001b[0mrlen\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFastaIndexingError\u001b[0m: Line length of fasta file is not consistent! Inconsistent line found in >sacCer3_ct_UserTrack_3545_YAL055W_SC_Transcript_00009455_Miura_2006 at line 21."
     ]
    }
   ],
   "source": [
    "five_prime_idx = pyfaidx.Fasta(five_prime_fasta)\n",
    "three_prime_idx = pyfaidx.Fasta(three_prime_fasta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove random # lines from fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "five_prime_cleaned_fasta = '/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/SGD_all_ORFs/SGD_all_ORFs_5prime_UTRs_cleaned.fsa'\n",
    "three_prime_cleaned_fasta = '/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/SGD_all_ORFs/SGD_all_ORFs_3prime_UTRs_cleaned.fsa'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "lines = []\n",
    "with open(five_prime_fasta) as fh:\n",
    "    for line in fh:\n",
    "        if line[0]!=\"#\":\n",
    "            lines.append(line)\n",
    "with open(five_prime_cleaned_fasta, 'w') as fh:\n",
    "    fh.writelines(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "lines = []\n",
    "with open(three_prime_fasta) as fh:\n",
    "    for line in fh:\n",
    "        if line[0]!=\"#\":\n",
    "            lines.append(line)\n",
    "with open(three_prime_cleaned_fasta, 'w') as fh:\n",
    "    fh.writelines(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "five_prime_idx = pyfaidx.Fasta(five_prime_cleaned_fasta)\n",
    "three_prime_idx = pyfaidx.Fasta(three_prime_cleaned_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "five_prime_records = []\n",
    "for record in SeqIO.parse(five_prime_cleaned_fasta, \"fasta\"):\n",
    "    id = record.id\n",
    "    description = record.description.replace(id, '').strip()\n",
    "    chrom_start_stop = description.split(\" 5'pad\")[0].replace('range=', '')\n",
    "    chrom = chrom_start_stop.split(':')[0].replace('chr', '')\n",
    "    start,stop = chrom_start_stop.split('-')\n",
    "    stop = int(stop)\n",
    "    start = start.split(':')[1]\n",
    "    start = int(start)\n",
    "    strand = description.split('strand=')[1][0]\n",
    "    five_pad = description.split(\"5'pad\")[1].split(\" \")[0]\n",
    "    three_pad = description.split(\"3'pad\")[1].split(\" \")[0]\n",
    "    \n",
    "    gene_name = id.split('_')[4]\n",
    "    source = ('_').join(id.split('_')[8:10])\n",
    "    five_prime_records.append((chrom, source, 'five_prime_utr', start, stop, '.', strand, '.',  'gene_id \"{}\"; transcript_id \"{}_mRNA\"; gene_source \"{}\"; gene_biotype \"protein_coding\"; transcript_biotype \"protein_coding\";'.format(gene_name, gene_name, source)))\n",
    "    five_prime_records.append((chrom, source, 'exon', start, stop, '.', strand, '.',  'gene_id \"{}\"; transcript_id \"{}_mRNA\"; gene_source \"{}\"; gene_biotype \"protein_coding\"; transcript_biotype \"protein_coding\";'.format(gene_name, gene_name, source)))\n",
    "five_prime_records_df = pd.DataFrame(five_prime_records).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "three_prime_records = []\n",
    "for record in SeqIO.parse(three_prime_cleaned_fasta, \"fasta\"):\n",
    "    id = record.id\n",
    "    description = record.description.replace(id, '').strip()\n",
    "    chrom_start_stop = description.split(\" 5'pad\")[0].replace('range=', '')\n",
    "    chrom = chrom_start_stop.split(':')[0].replace('chr', '')\n",
    "    start,stop = chrom_start_stop.split('-')\n",
    "    stop = int(stop)\n",
    "    start = start.split(':')[1]\n",
    "    start = int(start)\n",
    "    strand = description.split('strand=')[1][0]\n",
    "    three_pad = description.split(\"5'pad\")[1].split(\" \")[0]\n",
    "    three_pad = description.split(\"3'pad\")[1].split(\" \")[0]\n",
    "    \n",
    "    gene_name = id.split('_')[4]\n",
    "    source = ('_').join(id.split('_')[8:10])\n",
    "    three_prime_records.append((chrom, source, 'three_prime_utr', start, stop, '.', strand, '.',  'gene_id \"{}\"; transcript_id \"{}_mRNA\"; gene_source \"{}\"; gene_biotype \"protein_coding\"; transcript_biotype \"protein_coding\";'.format(gene_name, gene_name, source)))\n",
    "    three_prime_records.append((chrom, source, 'exon', start, stop, '.', strand, '.',  'gene_id \"{}\"; transcript_id \"{}_mRNA\"; gene_source \"{}\"; gene_biotype \"protein_coding\"; transcript_biotype \"protein_coding\";'.format(gene_name, gene_name, source)))\n",
    "three_prime_records_df = pd.DataFrame(three_prime_records).drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>I</td>\n",
       "      <td>Nagalakshmi_2008</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>36304</td>\n",
       "      <td>36420</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>I</td>\n",
       "      <td>Nagalakshmi_2008</td>\n",
       "      <td>exon</td>\n",
       "      <td>36304</td>\n",
       "      <td>36420</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>I</td>\n",
       "      <td>Yassour_2009</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>36304</td>\n",
       "      <td>36408</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>I</td>\n",
       "      <td>Yassour_2009</td>\n",
       "      <td>exon</td>\n",
       "      <td>36304</td>\n",
       "      <td>36408</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>I</td>\n",
       "      <td>Xu_2009</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>36304</td>\n",
       "      <td>36393</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35795</td>\n",
       "      <td>XIII</td>\n",
       "      <td>Nagalakshmi_2008</td>\n",
       "      <td>exon</td>\n",
       "      <td>905836</td>\n",
       "      <td>905904</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YMR316W\"; transcript_id \"YMR316W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35798</td>\n",
       "      <td>XIII</td>\n",
       "      <td>Miura_2006</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>910966</td>\n",
       "      <td>911061</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35799</td>\n",
       "      <td>XIII</td>\n",
       "      <td>Miura_2006</td>\n",
       "      <td>exon</td>\n",
       "      <td>910966</td>\n",
       "      <td>911061</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35800</td>\n",
       "      <td>XIII</td>\n",
       "      <td>Xu_2009</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>910966</td>\n",
       "      <td>911061</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>35801</td>\n",
       "      <td>XIII</td>\n",
       "      <td>Xu_2009</td>\n",
       "      <td>exon</td>\n",
       "      <td>910966</td>\n",
       "      <td>911061</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15856 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                 1                2       3       4  5  6  7  \\\n",
       "0         I  Nagalakshmi_2008  three_prime_utr   36304   36420  .  +  .   \n",
       "1         I  Nagalakshmi_2008             exon   36304   36420  .  +  .   \n",
       "4         I      Yassour_2009  three_prime_utr   36304   36408  .  +  .   \n",
       "5         I      Yassour_2009             exon   36304   36408  .  +  .   \n",
       "6         I           Xu_2009  three_prime_utr   36304   36393  .  +  .   \n",
       "...     ...               ...              ...     ...     ... .. .. ..   \n",
       "35795  XIII  Nagalakshmi_2008             exon  905836  905904  .  +  .   \n",
       "35798  XIII        Miura_2006  three_prime_utr  910966  911061  .  -  .   \n",
       "35799  XIII        Miura_2006             exon  910966  911061  .  -  .   \n",
       "35800  XIII           Xu_2009  three_prime_utr  910966  911061  .  -  .   \n",
       "35801  XIII           Xu_2009             exon  910966  911061  .  -  .   \n",
       "\n",
       "                                                       8  \n",
       "0      gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...  \n",
       "1      gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...  \n",
       "4      gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...  \n",
       "5      gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...  \n",
       "6      gene_id \"YAL060W\"; transcript_id \"YAL060W_mRNA...  \n",
       "...                                                  ...  \n",
       "35795  gene_id \"YMR316W\"; transcript_id \"YMR316W_mRNA...  \n",
       "35798  gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...  \n",
       "35799  gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...  \n",
       "35800  gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...  \n",
       "35801  gene_id \"YMR318C\"; transcript_id \"YMR318C_mRNA...  \n",
       "\n",
       "[15856 rows x 9 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "three_prime_records_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Saccharomyces_cerevisiae.R64-1-1.96.gtf'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.basename(original_gtf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge the GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_gtf = '/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/Saccharomyces_cerevisiae.R64-1-1.96.gtf'\n",
    "original_gtf_lines = []\n",
    "with open(original_gtf) as fh:\n",
    "    for line in fh:\n",
    "        original_gtf_lines.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>IV</td>\n",
       "      <td>sgd</td>\n",
       "      <td>gene</td>\n",
       "      <td>1802</td>\n",
       "      <td>2953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YDL248W\"; gene_source \"sgd\"; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>IV</td>\n",
       "      <td>sgd</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1802</td>\n",
       "      <td>2953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>IV</td>\n",
       "      <td>sgd</td>\n",
       "      <td>exon</td>\n",
       "      <td>1802</td>\n",
       "      <td>2953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>IV</td>\n",
       "      <td>sgd</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1802</td>\n",
       "      <td>2950</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>IV</td>\n",
       "      <td>sgd</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1802</td>\n",
       "      <td>1804</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0    1            2     3     4  5  6  7  \\\n",
       "0  IV  sgd         gene  1802  2953  .  +  .   \n",
       "1  IV  sgd   transcript  1802  2953  .  +  .   \n",
       "2  IV  sgd         exon  1802  2953  .  +  .   \n",
       "3  IV  sgd          CDS  1802  2950  .  +  0   \n",
       "4  IV  sgd  start_codon  1802  1804  .  +  0   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"YDL248W\"; gene_source \"sgd\"; gene_bio...  \n",
       "1  gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...  \n",
       "2  gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...  \n",
       "3  gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...  \n",
       "4  gene_id \"YDL248W\"; transcript_id \"YDL248W_mRNA...  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "original_gtf_df = pd.read_csv(original_gtf, sep=\"\\t\", skiprows=5, header=None)\n",
    "original_gtf_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df_combined = pd.concat([original_gtf_df, five_prime_records_df, three_prime_records_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df_combined = gtf_df_combined.sort_values(by=[0, 3,4,2, 6])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df_combined.to_csv(\"/panfs/qcb-panasas/skchoudh/genomes/R64-1-1/annotation/Saccharomyces_cerevisiae.R64-1-1.96.combined.UTR5UTR3.gtf\", sep=\"\\t\", index=False, header=False) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>40567</td>\n",
       "      <td>I</td>\n",
       "      <td>sgd</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>335</td>\n",
       "      <td>337</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40566</td>\n",
       "      <td>I</td>\n",
       "      <td>sgd</td>\n",
       "      <td>CDS</td>\n",
       "      <td>335</td>\n",
       "      <td>646</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40565</td>\n",
       "      <td>I</td>\n",
       "      <td>sgd</td>\n",
       "      <td>exon</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40563</td>\n",
       "      <td>I</td>\n",
       "      <td>sgd</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL069W\"; gene_source \"sgd\"; gene_bio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>40564</td>\n",
       "      <td>I</td>\n",
       "      <td>sgd</td>\n",
       "      <td>transcript</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0    1            2    3    4  5  6  7  \\\n",
       "40567  I  sgd  start_codon  335  337  .  +  0   \n",
       "40566  I  sgd          CDS  335  646  .  +  0   \n",
       "40565  I  sgd         exon  335  649  .  +  .   \n",
       "40563  I  sgd         gene  335  649  .  +  .   \n",
       "40564  I  sgd   transcript  335  649  .  +  .   \n",
       "\n",
       "                                                       8  \n",
       "40567  gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...  \n",
       "40566  gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...  \n",
       "40565  gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...  \n",
       "40563  gene_id \"YAL069W\"; gene_source \"sgd\"; gene_bio...  \n",
       "40564  gene_id \"YAL069W\"; transcript_id \"YAL069W_mRNA...  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df_combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (riboraptor)",
   "language": "python",
   "name": "riboraptor"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
